{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "10b51472",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import math\n",
    "\n",
    "class FastAreader:\n",
    "    def __init__(self, fname=''):\n",
    "        \"\"\"contructor: saves attribute fname\"\"\"\n",
    "        self.fname = fname\n",
    "        self.fileH = None\n",
    "        \n",
    "    def doOpen(self):\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "    \n",
    "    def readFasta(self):\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        with self.doOpen() as self.fileH:\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "            # skip to first fasta header\n",
    "            line = self.fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = self.fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "            \n",
    "            for line in self.fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        yield header, sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "40c93999",
   "metadata": {},
   "outputs": [],
   "source": [
    "fr = FastAreader(\"data/Ecoli-UMN026 (2).txt\")\n",
    "dna = list(fr.readFasta())[0][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "91f6405c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CommandLine():\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond.\n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "    '''\n",
    "    \n",
    "    def __init__(self, inOpts=None):\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "        Implements a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "        description='Program prolog - a brief description of what this thing does',\n",
    "        epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "        add_help=True, # default is True\n",
    "        prefix_chars='-',\n",
    "        usage='%(prog)s [options] -option1[default] <input >output'\n",
    "        )\n",
    "        \n",
    "        self.parser.add_argument('-l', '--minMotif', nargs='?', type=int, default=1, action='store',\n",
    "        help='min kMer size ')\n",
    "        \n",
    "        self.parser.add_argument('-m', '--maxMotif', nargs='?', type=int, default=8, action='store',\n",
    "        help='max kMer size ')\n",
    "        \n",
    "        self.parser.add_argument('-c', '--cutoff', nargs='?', type=float, default=.01, action='store',\n",
    "        help='Zscore cutoff')\n",
    "        \n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n",
    "        \n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5ff7f984",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genome:\n",
    "    \"\"\" \"\"\"\n",
    "    def __init__(self, minK, maxK):\n",
    "        \"\"\"Construct the object to count and analyze kMer frequencies\"\"\"\n",
    "        self.mink = minK\n",
    "        self.maxk = maxK\n",
    "        \n",
    "    def reverse_complement(self, dna):\n",
    "        basepairs = {\"A\":\"T\", \"G\":\"C\", \"T\":\"A\", \"C\":\"G\"}\n",
    "        return \"\".join([basepairs[base] for base in dna])[::-1]\n",
    "        \n",
    "    def E(self, K, counts, N):\n",
    "        ex = 1\n",
    "        for i in range(len(K)):\n",
    "            if i == 0:\n",
    "                ex *= (counts[K[i]]/N)\n",
    "            else:\n",
    "                kmer = K[:i+1]\n",
    "                ex *= (counts[kmer]/counts[kmer[:-1]])\n",
    "        return ex\n",
    "        \n",
    "    def count(self, genome):\n",
    "        counts = {}\n",
    "        bases = {'A', 'C', 'G', 'T'}\n",
    "        for i in range(1, self.maxk+1):\n",
    "            for k in range(len(genome)-i+1):\n",
    "                kmer = genome[k:k+i]\n",
    "                if set(kmer).issubset(bases):\n",
    "                    counts[kmer] = counts.get(kmer, 0) + 1\n",
    "        return counts\n",
    "    \n",
    "    def z_score(self, s, mu, sd):\n",
    "        try:\n",
    "            return (s - mu)/sd\n",
    "        except ZeroDivisionError:\n",
    "            return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2582e912",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main (inFile=None, options = None):\n",
    "    ''' Setup necessary objects, read data and print the final report.'''\n",
    "    cl = CommandLine(options) # setup the command line\n",
    "    sourceReader = FastAreader(inFile) # setup the Fasta reader Object\n",
    "    thisGenome = Genome(cl.args.minMotif, cl.args.maxMotif) # setup a Genome object\n",
    "    ###\n",
    "    # Give your Genome some data and tell it to do stuff\n",
    "    for Id, genome in sourceReader.readFasta():\n",
    "        counts = thisGenome.count(genome)\n",
    "        for kmer, s in counts.items():\n",
    "            if len(kmer) >= cl.args.minMotif:\n",
    "                rev_kmer = thisGenome.reverse_complement(kmer)\n",
    "                if (kmer == rev_kmer) or (rev_kmer not in counts.keys()):\n",
    "                    total_count = counts[kmer]\n",
    "                else:\n",
    "                    total_count = counts[kmer] + counts[rev_kmer]\n",
    "                    \n",
    "                n = len(genome) - 8\n",
    "                if (kmer == rev_kmer) or (rev_kmer not in counts.keys()):\n",
    "                    ex = (thisGenome.E(kmer, counts, n) + thisGenome.E(kmer, counts, n)) * (2*n)\n",
    "                else:\n",
    "                    ex = thisGenome.E(kmer, counts, n) + thisGenome.E(rev_kmer, counts, n) * (2*n)\n",
    "                p = ex/n\n",
    "                mu = n*p\n",
    "                sd = (mu * (1-p))**0.5\n",
    "                zscore = thisGenome.z_score(s, mu, sd)\n",
    "                #print(s, mu, sd, zscore)\n",
    "                if zscore < cl.args.cutoff:\n",
    "                    print(s, mu, sd, zscore)\n",
    "    ###\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    main(inFile=\"data/Ecoli-UMN026 (2).txt\", options=[\"--minMotif=3\", \"--maxMotif=8\", \"--cutoff=-4.\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60253265",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
