{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5660b29-ad41-4875-a8b6-956f2eaa36f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import Seq\n",
    "from Bio.Seq import translate\n",
    "from Bio.Data.CodonTable import standard_dna_table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b133cc4-61ad-4359-b4c1-9d8d51f037c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'F': ['TTT', 'TTC'],\n",
       " 'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
       " 'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
       " 'Y': ['TAT', 'TAC'],\n",
       " 'C': ['TGT', 'TGC'],\n",
       " 'W': ['TGG'],\n",
       " 'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
       " 'H': ['CAT', 'CAC'],\n",
       " 'Q': ['CAA', 'CAG'],\n",
       " 'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
       " 'I': ['ATT', 'ATC', 'ATA'],\n",
       " 'M': ['ATG'],\n",
       " 'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
       " 'N': ['AAT', 'AAC'],\n",
       " 'K': ['AAA', 'AAG'],\n",
       " 'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
       " 'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
       " 'D': ['GAT', 'GAC'],\n",
       " 'E': ['GAA', 'GAG'],\n",
       " 'G': ['GGT', 'GGC', 'GGA', 'GGG']}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rev_table = {}\n",
    "for codon, aa in standard_dna_table.forward_table.items():\n",
    "    rev_table.setdefault(aa, []).append(codon)\n",
    "\n",
    "rev_table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8b7095ce-d1e5-4d51-99d3-5ac03945c4fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_27525/3134801216.py:1: DtypeWarning: Columns (18,19,20,21,22,23) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  clinvar_df = pd.read_csv(\"clinvar_result.txt\", sep=\"\\t\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>...</th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Germline date last evaluated</th>\n",
       "      <th>Germline review status</th>\n",
       "      <th>Somatic clinical impact</th>\n",
       "      <th>Somatic clinical impact date last evaluated</th>\n",
       "      <th>Somatic clinical impact review status</th>\n",
       "      <th>Oncogenicity classification</th>\n",
       "      <th>Oncogenicity date last evaluated</th>\n",
       "      <th>Oncogenicity review status</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37/hg19 1q21.1(chr1:143134063-143284670)x3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000145208</td>\n",
       "      <td>1</td>\n",
       "      <td>143134063 - 143284670</td>\n",
       "      <td>1</td>\n",
       "      <td>16302 - 166909</td>\n",
       "      <td>145208</td>\n",
       "      <td>...</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Nov 30, 2010</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001005484.2(OR4F5):c.107A&gt;G (p.Glu36Gly)</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>E36G</td>\n",
       "      <td>not specified</td>\n",
       "      <td>VCV002205837</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>2205837</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Aug 23, 2021</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh37/hg19 1p36.11(chr1:26930197-26968251)x1</td>\n",
       "      <td>LOC129929833|LOC129929834|LOC129929835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000151336</td>\n",
       "      <td>1</td>\n",
       "      <td>26930197 - 26968251</td>\n",
       "      <td>1</td>\n",
       "      <td>91324 - 128277</td>\n",
       "      <td>151336</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Apr 30, 2011</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh38/hg38 1p36.33-35.1(chr1:99125-34026935)x3</td>\n",
       "      <td>A2ML1-AS2|A3GALT2|AADACL3|AADACL4|ACAP3|ACOT7|...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trisomy 12p</td>\n",
       "      <td>VCV002664871</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>99125 - 34026935</td>\n",
       "      <td>2664871</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Dec 12, 2023</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRCh38/hg38 1p36.33-36.23(chr1:629025-8537745)x1</td>\n",
       "      <td>LOC121967050|LOC121967051|LOC121967052|LOC1219...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000058242</td>\n",
       "      <td>1</td>\n",
       "      <td>564405 - 8597804</td>\n",
       "      <td>1</td>\n",
       "      <td>629025 - 8537745</td>\n",
       "      <td>58242</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Aug 12, 2011</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451376</th>\n",
       "      <td>NF1, 80-KB DEL</td>\n",
       "      <td>NF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Café-au-lait macules with pulmonary stenosis</td>\n",
       "      <td>VCV000000342</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>342</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Dec 1, 1992</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451377</th>\n",
       "      <td>NG_009018.1:g.246621_246622insN[320]</td>\n",
       "      <td>NF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neurofibromatosis, type 1</td>\n",
       "      <td>VCV000000332</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>332</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Oct 31, 1991</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451378</th>\n",
       "      <td>NM_004628.4(XPC):c.621_622ins83 (p.?)</td>\n",
       "      <td>XPC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xeroderma pigmentosum, group C</td>\n",
       "      <td>VCV000000254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>254</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Dec 1, 1993</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451379</th>\n",
       "      <td>NG_023416.2:g.(88604_176551)_(176714_576894)del</td>\n",
       "      <td>HPSE2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Urofacial syndrome type 1</td>\n",
       "      <td>VCV000000087</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jun 11, 2010</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451380</th>\n",
       "      <td>NM_021828.5:c.1099-4166_1320+840delins23</td>\n",
       "      <td>HPSE2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Urofacial syndrome type 1</td>\n",
       "      <td>VCV000000086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jun 11, 2010</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1451381 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Name  \\\n",
       "0          GRCh37/hg19 1q21.1(chr1:143134063-143284670)x3   \n",
       "1             NM_001005484.2(OR4F5):c.107A>G (p.Glu36Gly)   \n",
       "2           GRCh37/hg19 1p36.11(chr1:26930197-26968251)x1   \n",
       "3         GRCh38/hg38 1p36.33-35.1(chr1:99125-34026935)x3   \n",
       "4        GRCh38/hg38 1p36.33-36.23(chr1:629025-8537745)x1   \n",
       "...                                                   ...   \n",
       "1451376                                    NF1, 80-KB DEL   \n",
       "1451377              NG_009018.1:g.246621_246622insN[320]   \n",
       "1451378             NM_004628.4(XPC):c.621_622ins83 (p.?)   \n",
       "1451379   NG_023416.2:g.(88604_176551)_(176714_576894)del   \n",
       "1451380          NM_021828.5:c.1099-4166_1320+840delins23   \n",
       "\n",
       "                                                   Gene(s) Protein change  \\\n",
       "0                                                      NaN            NaN   \n",
       "1                                                    OR4F5           E36G   \n",
       "2                   LOC129929833|LOC129929834|LOC129929835            NaN   \n",
       "3        A2ML1-AS2|A3GALT2|AADACL3|AADACL4|ACAP3|ACOT7|...            NaN   \n",
       "4        LOC121967050|LOC121967051|LOC121967052|LOC1219...            NaN   \n",
       "...                                                    ...            ...   \n",
       "1451376                                                NF1            NaN   \n",
       "1451377                                                NF1            NaN   \n",
       "1451378                                                XPC            NaN   \n",
       "1451379                                              HPSE2            NaN   \n",
       "1451380                                              HPSE2            NaN   \n",
       "\n",
       "                                         Condition(s)     Accession  \\\n",
       "0                                           See cases  VCV000145208   \n",
       "1                                       not specified  VCV002205837   \n",
       "2                                           See cases  VCV000151336   \n",
       "3                                         Trisomy 12p  VCV002664871   \n",
       "4                                           See cases  VCV000058242   \n",
       "...                                               ...           ...   \n",
       "1451376  Café-au-lait macules with pulmonary stenosis  VCV000000342   \n",
       "1451377                     Neurofibromatosis, type 1  VCV000000332   \n",
       "1451378                Xeroderma pigmentosum, group C  VCV000000254   \n",
       "1451379                     Urofacial syndrome type 1  VCV000000087   \n",
       "1451380                     Urofacial syndrome type 1  VCV000000086   \n",
       "\n",
       "        GRCh37Chromosome         GRCh37Location GRCh38Chromosome  \\\n",
       "0                      1  143134063 - 143284670                1   \n",
       "1                      1                  69134                1   \n",
       "2                      1    26930197 - 26968251                1   \n",
       "3                    NaN                    NaN                1   \n",
       "4                      1       564405 - 8597804                1   \n",
       "...                  ...                    ...              ...   \n",
       "1451376              NaN                    NaN              NaN   \n",
       "1451377              NaN                    NaN              NaN   \n",
       "1451378              NaN                    NaN              NaN   \n",
       "1451379              NaN                    NaN              NaN   \n",
       "1451380              NaN                    NaN              NaN   \n",
       "\n",
       "           GRCh38Location  VariationID  ... Germline classification  \\\n",
       "0          16302 - 166909       145208  ...                  Benign   \n",
       "1                   69134      2205837  ...           Likely benign   \n",
       "2          91324 - 128277       151336  ...           Likely benign   \n",
       "3        99125 - 34026935      2664871  ...              Pathogenic   \n",
       "4        629025 - 8537745        58242  ...              Pathogenic   \n",
       "...                   ...          ...  ...                     ...   \n",
       "1451376               NaN          342  ...              Pathogenic   \n",
       "1451377               NaN          332  ...              Pathogenic   \n",
       "1451378               NaN          254  ...              Pathogenic   \n",
       "1451379               NaN           87  ...              Pathogenic   \n",
       "1451380               NaN           86  ...              Pathogenic   \n",
       "\n",
       "        Germline date last evaluated               Germline review status  \\\n",
       "0                       Nov 30, 2010       no assertion criteria provided   \n",
       "1                       Aug 23, 2021  criteria provided, single submitter   \n",
       "2                       Apr 30, 2011       no assertion criteria provided   \n",
       "3                       Dec 12, 2023  criteria provided, single submitter   \n",
       "4                       Aug 12, 2011  criteria provided, single submitter   \n",
       "...                              ...                                  ...   \n",
       "1451376                  Dec 1, 1992       no assertion criteria provided   \n",
       "1451377                 Oct 31, 1991       no assertion criteria provided   \n",
       "1451378                  Dec 1, 1993       no assertion criteria provided   \n",
       "1451379                 Jun 11, 2010       no assertion criteria provided   \n",
       "1451380                 Jun 11, 2010       no assertion criteria provided   \n",
       "\n",
       "        Somatic clinical impact Somatic clinical impact date last evaluated  \\\n",
       "0                           NaN                                         NaN   \n",
       "1                           NaN                                         NaN   \n",
       "2                           NaN                                         NaN   \n",
       "3                           NaN                                         NaN   \n",
       "4                           NaN                                         NaN   \n",
       "...                         ...                                         ...   \n",
       "1451376                     NaN                                         NaN   \n",
       "1451377                     NaN                                         NaN   \n",
       "1451378                     NaN                                         NaN   \n",
       "1451379                     NaN                                         NaN   \n",
       "1451380                     NaN                                         NaN   \n",
       "\n",
       "        Somatic clinical impact review status Oncogenicity classification  \\\n",
       "0                                         NaN                         NaN   \n",
       "1                                         NaN                         NaN   \n",
       "2                                         NaN                         NaN   \n",
       "3                                         NaN                         NaN   \n",
       "4                                         NaN                         NaN   \n",
       "...                                       ...                         ...   \n",
       "1451376                                   NaN                         NaN   \n",
       "1451377                                   NaN                         NaN   \n",
       "1451378                                   NaN                         NaN   \n",
       "1451379                                   NaN                         NaN   \n",
       "1451380                                   NaN                         NaN   \n",
       "\n",
       "        Oncogenicity date last evaluated Oncogenicity review status  \\\n",
       "0                                    NaN                        NaN   \n",
       "1                                    NaN                        NaN   \n",
       "2                                    NaN                        NaN   \n",
       "3                                    NaN                        NaN   \n",
       "4                                    NaN                        NaN   \n",
       "...                                  ...                        ...   \n",
       "1451376                              NaN                        NaN   \n",
       "1451377                              NaN                        NaN   \n",
       "1451378                              NaN                        NaN   \n",
       "1451379                              NaN                        NaN   \n",
       "1451380                              NaN                        NaN   \n",
       "\n",
       "        Unnamed: 24  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2               NaN  \n",
       "3               NaN  \n",
       "4               NaN  \n",
       "...             ...  \n",
       "1451376         NaN  \n",
       "1451377         NaN  \n",
       "1451378         NaN  \n",
       "1451379         NaN  \n",
       "1451380         NaN  \n",
       "\n",
       "[1451381 rows x 25 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_df = pd.read_csv(\"clinvar_result.txt\", sep=\"\\t\")\n",
    "clinvar_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "be0544e8-4278-4d94-bbad-4a03a9aadfbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>...</th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Germline date last evaluated</th>\n",
       "      <th>Germline review status</th>\n",
       "      <th>Somatic clinical impact</th>\n",
       "      <th>Somatic clinical impact date last evaluated</th>\n",
       "      <th>Somatic clinical impact review status</th>\n",
       "      <th>Oncogenicity classification</th>\n",
       "      <th>Oncogenicity date last evaluated</th>\n",
       "      <th>Oncogenicity review status</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRCh37/hg19 1q21.1(chr1:143134063-143284670)x3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000145208</td>\n",
       "      <td>1</td>\n",
       "      <td>143134063 - 143284670</td>\n",
       "      <td>1</td>\n",
       "      <td>16302 - 166909</td>\n",
       "      <td>145208</td>\n",
       "      <td>...</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Nov 30, 2010</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001005484.2(OR4F5):c.107A&gt;G (p.Glu36Gly)</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>E36G</td>\n",
       "      <td>not specified</td>\n",
       "      <td>VCV002205837</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>2205837</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Aug 23, 2021</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRCh37/hg19 1p36.11(chr1:26930197-26968251)x1</td>\n",
       "      <td>LOC129929833|LOC129929834|LOC129929835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000151336</td>\n",
       "      <td>1</td>\n",
       "      <td>26930197 - 26968251</td>\n",
       "      <td>1</td>\n",
       "      <td>91324 - 128277</td>\n",
       "      <td>151336</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Apr 30, 2011</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRCh38/hg38 1p36.33-35.1(chr1:99125-34026935)x3</td>\n",
       "      <td>A2ML1-AS2|A3GALT2|AADACL3|AADACL4|ACAP3|ACOT7|...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trisomy 12p</td>\n",
       "      <td>VCV002664871</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>99125 - 34026935</td>\n",
       "      <td>2664871</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Dec 12, 2023</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRCh38/hg38 1p36.33-36.23(chr1:629025-8537745)x1</td>\n",
       "      <td>LOC121967050|LOC121967051|LOC121967052|LOC1219...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>See cases</td>\n",
       "      <td>VCV000058242</td>\n",
       "      <td>1</td>\n",
       "      <td>564405 - 8597804</td>\n",
       "      <td>1</td>\n",
       "      <td>629025 - 8537745</td>\n",
       "      <td>58242</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Aug 12, 2011</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Name  \\\n",
       "0    GRCh37/hg19 1q21.1(chr1:143134063-143284670)x3   \n",
       "1       NM_001005484.2(OR4F5):c.107A>G (p.Glu36Gly)   \n",
       "2     GRCh37/hg19 1p36.11(chr1:26930197-26968251)x1   \n",
       "3   GRCh38/hg38 1p36.33-35.1(chr1:99125-34026935)x3   \n",
       "4  GRCh38/hg38 1p36.33-36.23(chr1:629025-8537745)x1   \n",
       "\n",
       "                                             Gene(s) Protein change  \\\n",
       "0                                                NaN            NaN   \n",
       "1                                              OR4F5           E36G   \n",
       "2             LOC129929833|LOC129929834|LOC129929835            NaN   \n",
       "3  A2ML1-AS2|A3GALT2|AADACL3|AADACL4|ACAP3|ACOT7|...            NaN   \n",
       "4  LOC121967050|LOC121967051|LOC121967052|LOC1219...            NaN   \n",
       "\n",
       "    Condition(s)     Accession GRCh37Chromosome         GRCh37Location  \\\n",
       "0      See cases  VCV000145208                1  143134063 - 143284670   \n",
       "1  not specified  VCV002205837                1                  69134   \n",
       "2      See cases  VCV000151336                1    26930197 - 26968251   \n",
       "3    Trisomy 12p  VCV002664871              NaN                    NaN   \n",
       "4      See cases  VCV000058242                1       564405 - 8597804   \n",
       "\n",
       "  GRCh38Chromosome    GRCh38Location  VariationID  ...  \\\n",
       "0                1    16302 - 166909       145208  ...   \n",
       "1                1             69134      2205837  ...   \n",
       "2                1    91324 - 128277       151336  ...   \n",
       "3                1  99125 - 34026935      2664871  ...   \n",
       "4                1  629025 - 8537745        58242  ...   \n",
       "\n",
       "  Germline classification Germline date last evaluated  \\\n",
       "0                  Benign                 Nov 30, 2010   \n",
       "1           Likely benign                 Aug 23, 2021   \n",
       "2           Likely benign                 Apr 30, 2011   \n",
       "3              Pathogenic                 Dec 12, 2023   \n",
       "4              Pathogenic                 Aug 12, 2011   \n",
       "\n",
       "                Germline review status Somatic clinical impact  \\\n",
       "0       no assertion criteria provided                     NaN   \n",
       "1  criteria provided, single submitter                     NaN   \n",
       "2       no assertion criteria provided                     NaN   \n",
       "3  criteria provided, single submitter                     NaN   \n",
       "4  criteria provided, single submitter                     NaN   \n",
       "\n",
       "  Somatic clinical impact date last evaluated  \\\n",
       "0                                         NaN   \n",
       "1                                         NaN   \n",
       "2                                         NaN   \n",
       "3                                         NaN   \n",
       "4                                         NaN   \n",
       "\n",
       "  Somatic clinical impact review status Oncogenicity classification  \\\n",
       "0                                   NaN                         NaN   \n",
       "1                                   NaN                         NaN   \n",
       "2                                   NaN                         NaN   \n",
       "3                                   NaN                         NaN   \n",
       "4                                   NaN                         NaN   \n",
       "\n",
       "  Oncogenicity date last evaluated Oncogenicity review status Unnamed: 24  \n",
       "0                              NaN                        NaN         NaN  \n",
       "1                              NaN                        NaN         NaN  \n",
       "2                              NaN                        NaN         NaN  \n",
       "3                              NaN                        NaN         NaN  \n",
       "4                              NaN                        NaN         NaN  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4cfc59e-2758-43b1-9b7e-ab0d09d26c6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession',\n",
       "       'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome',\n",
       "       'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID',\n",
       "       'Canonical SPDI', 'Variant type', 'Molecular consequence',\n",
       "       'Germline classification', 'Germline date last evaluated',\n",
       "       'Germline review status', 'Somatic clinical impact',\n",
       "       'Somatic clinical impact date last evaluated',\n",
       "       'Somatic clinical impact review status', 'Oncogenicity classification',\n",
       "       'Oncogenicity date last evaluated', 'Oncogenicity review status',\n",
       "       'Unnamed: 24'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3037603d-37f9-4dc9-8d3b-af0f5631916b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Name                                           NM_001005484.2(OR4F5):c.107A>G (p.Glu36Gly)\n",
       "Gene(s)                                                                              OR4F5\n",
       "Protein change                                                                        E36G\n",
       "Condition(s)                                                                 not specified\n",
       "Accession                                                                     VCV002205837\n",
       "GRCh37Chromosome                                                                         1\n",
       "GRCh37Location                                                                       69134\n",
       "GRCh38Chromosome                                                                         1\n",
       "GRCh38Location                                                                       69134\n",
       "VariationID                                                                        2205837\n",
       "AlleleID(s)                                                                        2193183\n",
       "dbSNP ID                                                                               NaN\n",
       "Canonical SPDI                                                      NC_000001.11:69133:A:G\n",
       "Variant type                                                     single nucleotide variant\n",
       "Molecular consequence                                                     missense variant\n",
       "Germline classification                                                      Likely benign\n",
       "Germline date last evaluated                                                  Aug 23, 2021\n",
       "Germline review status                                 criteria provided, single submitter\n",
       "Somatic clinical impact                                                                NaN\n",
       "Somatic clinical impact date last evaluated                                            NaN\n",
       "Somatic clinical impact review status                                                  NaN\n",
       "Oncogenicity classification                                                            NaN\n",
       "Oncogenicity date last evaluated                                                       NaN\n",
       "Oncogenicity review status                                                             NaN\n",
       "Unnamed: 24                                                                            NaN\n",
       "Name: 1, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_df.loc[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7183d7ee-1798-4b64-9eb0-73fd0d5b8b82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['copy number gain', 'single nucleotide variant',\n",
       "       'copy number loss', 'Deletion', 'Microsatellite', 'Indel',\n",
       "       'Duplication', 'Inversion', 'Insertion', 'Haplotype', 'Variation',\n",
       "       'Complex', 'fusion', 'Translocation', 'CompoundHeterozygote',\n",
       "       'Phase unknown', 'Diplotype', 'protein only',\n",
       "       'Distinct chromosomes'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_df['Variant type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f0faa506-de4a-42d4-8f88-f9fb1cb7b197",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Variant type\n",
       "single nucleotide variant    1225608\n",
       "Deletion                      115943\n",
       "Duplication                    46917\n",
       "Microsatellite                 22187\n",
       "copy number loss               12354\n",
       "copy number gain                9590\n",
       "Insertion                       9572\n",
       "Indel                           7303\n",
       "Haplotype                        477\n",
       "Inversion                        477\n",
       "Variation                        436\n",
       "CompoundHeterozygote             247\n",
       "Translocation                    143\n",
       "protein only                      74\n",
       "Complex                           41\n",
       "fusion                             5\n",
       "Phase unknown                      5\n",
       "Diplotype                          1\n",
       "Distinct chromosomes               1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_df['Variant type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b85821c9-dfad-42b2-9bb6-0e1f1489823d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>...</th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Germline date last evaluated</th>\n",
       "      <th>Germline review status</th>\n",
       "      <th>Somatic clinical impact</th>\n",
       "      <th>Somatic clinical impact date last evaluated</th>\n",
       "      <th>Somatic clinical impact review status</th>\n",
       "      <th>Oncogenicity classification</th>\n",
       "      <th>Oncogenicity date last evaluated</th>\n",
       "      <th>Oncogenicity review status</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001005484.2(OR4F5):c.107A&gt;G (p.Glu36Gly)</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>E36G</td>\n",
       "      <td>not specified</td>\n",
       "      <td>VCV002205837</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>2205837</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Aug 23, 2021</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.552C&gt;T (p.Ile184=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001543320</td>\n",
       "      <td>1</td>\n",
       "      <td>861336</td>\n",
       "      <td>1</td>\n",
       "      <td>925956</td>\n",
       "      <td>1543320</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Nov 20, 2023</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.565C&gt;T (p.Pro189Ser)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>P10S, P189S</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001648427</td>\n",
       "      <td>1</td>\n",
       "      <td>861349</td>\n",
       "      <td>1</td>\n",
       "      <td>925969</td>\n",
       "      <td>1648427</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Dec 9, 2023</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.576C&gt;T (p.Cys192=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001936488</td>\n",
       "      <td>1</td>\n",
       "      <td>861360</td>\n",
       "      <td>1</td>\n",
       "      <td>925980</td>\n",
       "      <td>1936488</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Mar 29, 2021</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.582C&gt;T (p.Cys194=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001568423</td>\n",
       "      <td>1</td>\n",
       "      <td>861366</td>\n",
       "      <td>1</td>\n",
       "      <td>925986</td>\n",
       "      <td>1568423</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Jul 5, 2022</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451320</th>\n",
       "      <td>MTHFR, 1081C-T</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homocystinuria due to methylene tetrahydrofola...</td>\n",
       "      <td>VCV000003526</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3526</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Apr 24, 2000</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451331</th>\n",
       "      <td>NPC1, IVS16, G-A, -82</td>\n",
       "      <td>NPC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Niemann-Pick disease, type C1</td>\n",
       "      <td>VCV000002975</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2975</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jul 1, 2001</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451340</th>\n",
       "      <td>NIPBL, IVS44DS, A-G, +4</td>\n",
       "      <td>NIPBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cornelia de Lange syndrome 1</td>\n",
       "      <td>VCV000002149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2149</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Oct 1, 2004</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451349</th>\n",
       "      <td>A114N</td>\n",
       "      <td>NIPAL4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Autosomal recessive congenital ichthyosis 6</td>\n",
       "      <td>VCV000001730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1730</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Oct 15, 2004</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451375</th>\n",
       "      <td>NF1, IVS34, G-A, +18</td>\n",
       "      <td>NF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Juvenile myelomonocytic leukemia</td>\n",
       "      <td>VCV000000351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>351</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jun 12, 1997</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1225608 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Name Gene(s) Protein change  \\\n",
       "1          NM_001005484.2(OR4F5):c.107A>G (p.Glu36Gly)   OR4F5           E36G   \n",
       "70         NM_001385641.1(SAMD11):c.552C>T (p.Ile184=)  SAMD11            NaN   \n",
       "71       NM_001385641.1(SAMD11):c.565C>T (p.Pro189Ser)  SAMD11    P10S, P189S   \n",
       "72         NM_001385641.1(SAMD11):c.576C>T (p.Cys192=)  SAMD11            NaN   \n",
       "73         NM_001385641.1(SAMD11):c.582C>T (p.Cys194=)  SAMD11            NaN   \n",
       "...                                                ...     ...            ...   \n",
       "1451320                                 MTHFR, 1081C-T   MTHFR            NaN   \n",
       "1451331                          NPC1, IVS16, G-A, -82    NPC1            NaN   \n",
       "1451340                        NIPBL, IVS44DS, A-G, +4   NIPBL            NaN   \n",
       "1451349                                          A114N  NIPAL4            NaN   \n",
       "1451375                           NF1, IVS34, G-A, +18     NF1            NaN   \n",
       "\n",
       "                                              Condition(s)     Accession  \\\n",
       "1                                            not specified  VCV002205837   \n",
       "70                                            not provided  VCV001543320   \n",
       "71                                            not provided  VCV001648427   \n",
       "72                                            not provided  VCV001936488   \n",
       "73                                            not provided  VCV001568423   \n",
       "...                                                    ...           ...   \n",
       "1451320  Homocystinuria due to methylene tetrahydrofola...  VCV000003526   \n",
       "1451331                      Niemann-Pick disease, type C1  VCV000002975   \n",
       "1451340                       Cornelia de Lange syndrome 1  VCV000002149   \n",
       "1451349        Autosomal recessive congenital ichthyosis 6  VCV000001730   \n",
       "1451375                   Juvenile myelomonocytic leukemia  VCV000000351   \n",
       "\n",
       "        GRCh37Chromosome GRCh37Location GRCh38Chromosome GRCh38Location  \\\n",
       "1                      1          69134                1          69134   \n",
       "70                     1         861336                1         925956   \n",
       "71                     1         861349                1         925969   \n",
       "72                     1         861360                1         925980   \n",
       "73                     1         861366                1         925986   \n",
       "...                  ...            ...              ...            ...   \n",
       "1451320              NaN            NaN              NaN            NaN   \n",
       "1451331              NaN            NaN              NaN            NaN   \n",
       "1451340              NaN            NaN              NaN            NaN   \n",
       "1451349              NaN            NaN              NaN            NaN   \n",
       "1451375              NaN            NaN              NaN            NaN   \n",
       "\n",
       "         VariationID  ... Germline classification  \\\n",
       "1            2205837  ...           Likely benign   \n",
       "70           1543320  ...           Likely benign   \n",
       "71           1648427  ...           Likely benign   \n",
       "72           1936488  ...           Likely benign   \n",
       "73           1568423  ...           Likely benign   \n",
       "...              ...  ...                     ...   \n",
       "1451320         3526  ...              Pathogenic   \n",
       "1451331         2975  ...              Pathogenic   \n",
       "1451340         2149  ...              Pathogenic   \n",
       "1451349         1730  ...              Pathogenic   \n",
       "1451375          351  ...              Pathogenic   \n",
       "\n",
       "        Germline date last evaluated               Germline review status  \\\n",
       "1                       Aug 23, 2021  criteria provided, single submitter   \n",
       "70                      Nov 20, 2023  criteria provided, single submitter   \n",
       "71                       Dec 9, 2023  criteria provided, single submitter   \n",
       "72                      Mar 29, 2021  criteria provided, single submitter   \n",
       "73                       Jul 5, 2022  criteria provided, single submitter   \n",
       "...                              ...                                  ...   \n",
       "1451320                 Apr 24, 2000       no assertion criteria provided   \n",
       "1451331                  Jul 1, 2001       no assertion criteria provided   \n",
       "1451340                  Oct 1, 2004       no assertion criteria provided   \n",
       "1451349                 Oct 15, 2004       no assertion criteria provided   \n",
       "1451375                 Jun 12, 1997       no assertion criteria provided   \n",
       "\n",
       "        Somatic clinical impact Somatic clinical impact date last evaluated  \\\n",
       "1                           NaN                                         NaN   \n",
       "70                          NaN                                         NaN   \n",
       "71                          NaN                                         NaN   \n",
       "72                          NaN                                         NaN   \n",
       "73                          NaN                                         NaN   \n",
       "...                         ...                                         ...   \n",
       "1451320                     NaN                                         NaN   \n",
       "1451331                     NaN                                         NaN   \n",
       "1451340                     NaN                                         NaN   \n",
       "1451349                     NaN                                         NaN   \n",
       "1451375                     NaN                                         NaN   \n",
       "\n",
       "        Somatic clinical impact review status Oncogenicity classification  \\\n",
       "1                                         NaN                         NaN   \n",
       "70                                        NaN                         NaN   \n",
       "71                                        NaN                         NaN   \n",
       "72                                        NaN                         NaN   \n",
       "73                                        NaN                         NaN   \n",
       "...                                       ...                         ...   \n",
       "1451320                                   NaN                         NaN   \n",
       "1451331                                   NaN                         NaN   \n",
       "1451340                                   NaN                         NaN   \n",
       "1451349                                   NaN                         NaN   \n",
       "1451375                                   NaN                         NaN   \n",
       "\n",
       "        Oncogenicity date last evaluated Oncogenicity review status  \\\n",
       "1                                    NaN                        NaN   \n",
       "70                                   NaN                        NaN   \n",
       "71                                   NaN                        NaN   \n",
       "72                                   NaN                        NaN   \n",
       "73                                   NaN                        NaN   \n",
       "...                                  ...                        ...   \n",
       "1451320                              NaN                        NaN   \n",
       "1451331                              NaN                        NaN   \n",
       "1451340                              NaN                        NaN   \n",
       "1451349                              NaN                        NaN   \n",
       "1451375                              NaN                        NaN   \n",
       "\n",
       "        Unnamed: 24  \n",
       "1               NaN  \n",
       "70              NaN  \n",
       "71              NaN  \n",
       "72              NaN  \n",
       "73              NaN  \n",
       "...             ...  \n",
       "1451320         NaN  \n",
       "1451331         NaN  \n",
       "1451340         NaN  \n",
       "1451349         NaN  \n",
       "1451375         NaN  \n",
       "\n",
       "[1225608 rows x 25 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snv_clinvar_df = clinvar_df[clinvar_df['Variant type'] == 'single nucleotide variant']\n",
    "snv_clinvar_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ae20c672-7dd7-4969-819e-ce003278a272",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_27525/3196568711.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['ClinVar:2205837' 'ClinVar:1543320' 'ClinVar:1648427' ... 'ClinVar:2149'\n",
      " 'ClinVar:1730' 'ClinVar:351']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  snv_clinvar_df.loc[:, 'VariationID'] = snv_clinvar_df['VariationID'].apply(lambda vid: f\"ClinVar:{vid}\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>...</th>\n",
       "      <th>Germline classification</th>\n",
       "      <th>Germline date last evaluated</th>\n",
       "      <th>Germline review status</th>\n",
       "      <th>Somatic clinical impact</th>\n",
       "      <th>Somatic clinical impact date last evaluated</th>\n",
       "      <th>Somatic clinical impact review status</th>\n",
       "      <th>Oncogenicity classification</th>\n",
       "      <th>Oncogenicity date last evaluated</th>\n",
       "      <th>Oncogenicity review status</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001005484.2(OR4F5):c.107A&gt;G (p.Glu36Gly)</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>E36G</td>\n",
       "      <td>not specified</td>\n",
       "      <td>VCV002205837</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>ClinVar:2205837</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Aug 23, 2021</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.552C&gt;T (p.Ile184=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001543320</td>\n",
       "      <td>1</td>\n",
       "      <td>861336</td>\n",
       "      <td>1</td>\n",
       "      <td>925956</td>\n",
       "      <td>ClinVar:1543320</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Nov 20, 2023</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.565C&gt;T (p.Pro189Ser)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>P10S, P189S</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001648427</td>\n",
       "      <td>1</td>\n",
       "      <td>861349</td>\n",
       "      <td>1</td>\n",
       "      <td>925969</td>\n",
       "      <td>ClinVar:1648427</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Dec 9, 2023</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.576C&gt;T (p.Cys192=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001936488</td>\n",
       "      <td>1</td>\n",
       "      <td>861360</td>\n",
       "      <td>1</td>\n",
       "      <td>925980</td>\n",
       "      <td>ClinVar:1936488</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Mar 29, 2021</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.582C&gt;T (p.Cys194=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001568423</td>\n",
       "      <td>1</td>\n",
       "      <td>861366</td>\n",
       "      <td>1</td>\n",
       "      <td>925986</td>\n",
       "      <td>ClinVar:1568423</td>\n",
       "      <td>...</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Jul 5, 2022</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451320</th>\n",
       "      <td>MTHFR, 1081C-T</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homocystinuria due to methylene tetrahydrofola...</td>\n",
       "      <td>VCV000003526</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:3526</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Apr 24, 2000</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451331</th>\n",
       "      <td>NPC1, IVS16, G-A, -82</td>\n",
       "      <td>NPC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Niemann-Pick disease, type C1</td>\n",
       "      <td>VCV000002975</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:2975</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jul 1, 2001</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451340</th>\n",
       "      <td>NIPBL, IVS44DS, A-G, +4</td>\n",
       "      <td>NIPBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cornelia de Lange syndrome 1</td>\n",
       "      <td>VCV000002149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:2149</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Oct 1, 2004</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451349</th>\n",
       "      <td>A114N</td>\n",
       "      <td>NIPAL4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Autosomal recessive congenital ichthyosis 6</td>\n",
       "      <td>VCV000001730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:1730</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Oct 15, 2004</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451375</th>\n",
       "      <td>NF1, IVS34, G-A, +18</td>\n",
       "      <td>NF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Juvenile myelomonocytic leukemia</td>\n",
       "      <td>VCV000000351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:351</td>\n",
       "      <td>...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Jun 12, 1997</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1225608 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Name Gene(s) Protein change  \\\n",
       "1          NM_001005484.2(OR4F5):c.107A>G (p.Glu36Gly)   OR4F5           E36G   \n",
       "70         NM_001385641.1(SAMD11):c.552C>T (p.Ile184=)  SAMD11            NaN   \n",
       "71       NM_001385641.1(SAMD11):c.565C>T (p.Pro189Ser)  SAMD11    P10S, P189S   \n",
       "72         NM_001385641.1(SAMD11):c.576C>T (p.Cys192=)  SAMD11            NaN   \n",
       "73         NM_001385641.1(SAMD11):c.582C>T (p.Cys194=)  SAMD11            NaN   \n",
       "...                                                ...     ...            ...   \n",
       "1451320                                 MTHFR, 1081C-T   MTHFR            NaN   \n",
       "1451331                          NPC1, IVS16, G-A, -82    NPC1            NaN   \n",
       "1451340                        NIPBL, IVS44DS, A-G, +4   NIPBL            NaN   \n",
       "1451349                                          A114N  NIPAL4            NaN   \n",
       "1451375                           NF1, IVS34, G-A, +18     NF1            NaN   \n",
       "\n",
       "                                              Condition(s)     Accession  \\\n",
       "1                                            not specified  VCV002205837   \n",
       "70                                            not provided  VCV001543320   \n",
       "71                                            not provided  VCV001648427   \n",
       "72                                            not provided  VCV001936488   \n",
       "73                                            not provided  VCV001568423   \n",
       "...                                                    ...           ...   \n",
       "1451320  Homocystinuria due to methylene tetrahydrofola...  VCV000003526   \n",
       "1451331                      Niemann-Pick disease, type C1  VCV000002975   \n",
       "1451340                       Cornelia de Lange syndrome 1  VCV000002149   \n",
       "1451349        Autosomal recessive congenital ichthyosis 6  VCV000001730   \n",
       "1451375                   Juvenile myelomonocytic leukemia  VCV000000351   \n",
       "\n",
       "        GRCh37Chromosome GRCh37Location GRCh38Chromosome GRCh38Location  \\\n",
       "1                      1          69134                1          69134   \n",
       "70                     1         861336                1         925956   \n",
       "71                     1         861349                1         925969   \n",
       "72                     1         861360                1         925980   \n",
       "73                     1         861366                1         925986   \n",
       "...                  ...            ...              ...            ...   \n",
       "1451320              NaN            NaN              NaN            NaN   \n",
       "1451331              NaN            NaN              NaN            NaN   \n",
       "1451340              NaN            NaN              NaN            NaN   \n",
       "1451349              NaN            NaN              NaN            NaN   \n",
       "1451375              NaN            NaN              NaN            NaN   \n",
       "\n",
       "             VariationID  ... Germline classification  \\\n",
       "1        ClinVar:2205837  ...           Likely benign   \n",
       "70       ClinVar:1543320  ...           Likely benign   \n",
       "71       ClinVar:1648427  ...           Likely benign   \n",
       "72       ClinVar:1936488  ...           Likely benign   \n",
       "73       ClinVar:1568423  ...           Likely benign   \n",
       "...                  ...  ...                     ...   \n",
       "1451320     ClinVar:3526  ...              Pathogenic   \n",
       "1451331     ClinVar:2975  ...              Pathogenic   \n",
       "1451340     ClinVar:2149  ...              Pathogenic   \n",
       "1451349     ClinVar:1730  ...              Pathogenic   \n",
       "1451375      ClinVar:351  ...              Pathogenic   \n",
       "\n",
       "        Germline date last evaluated               Germline review status  \\\n",
       "1                       Aug 23, 2021  criteria provided, single submitter   \n",
       "70                      Nov 20, 2023  criteria provided, single submitter   \n",
       "71                       Dec 9, 2023  criteria provided, single submitter   \n",
       "72                      Mar 29, 2021  criteria provided, single submitter   \n",
       "73                       Jul 5, 2022  criteria provided, single submitter   \n",
       "...                              ...                                  ...   \n",
       "1451320                 Apr 24, 2000       no assertion criteria provided   \n",
       "1451331                  Jul 1, 2001       no assertion criteria provided   \n",
       "1451340                  Oct 1, 2004       no assertion criteria provided   \n",
       "1451349                 Oct 15, 2004       no assertion criteria provided   \n",
       "1451375                 Jun 12, 1997       no assertion criteria provided   \n",
       "\n",
       "        Somatic clinical impact Somatic clinical impact date last evaluated  \\\n",
       "1                           NaN                                         NaN   \n",
       "70                          NaN                                         NaN   \n",
       "71                          NaN                                         NaN   \n",
       "72                          NaN                                         NaN   \n",
       "73                          NaN                                         NaN   \n",
       "...                         ...                                         ...   \n",
       "1451320                     NaN                                         NaN   \n",
       "1451331                     NaN                                         NaN   \n",
       "1451340                     NaN                                         NaN   \n",
       "1451349                     NaN                                         NaN   \n",
       "1451375                     NaN                                         NaN   \n",
       "\n",
       "        Somatic clinical impact review status Oncogenicity classification  \\\n",
       "1                                         NaN                         NaN   \n",
       "70                                        NaN                         NaN   \n",
       "71                                        NaN                         NaN   \n",
       "72                                        NaN                         NaN   \n",
       "73                                        NaN                         NaN   \n",
       "...                                       ...                         ...   \n",
       "1451320                                   NaN                         NaN   \n",
       "1451331                                   NaN                         NaN   \n",
       "1451340                                   NaN                         NaN   \n",
       "1451349                                   NaN                         NaN   \n",
       "1451375                                   NaN                         NaN   \n",
       "\n",
       "        Oncogenicity date last evaluated Oncogenicity review status  \\\n",
       "1                                    NaN                        NaN   \n",
       "70                                   NaN                        NaN   \n",
       "71                                   NaN                        NaN   \n",
       "72                                   NaN                        NaN   \n",
       "73                                   NaN                        NaN   \n",
       "...                                  ...                        ...   \n",
       "1451320                              NaN                        NaN   \n",
       "1451331                              NaN                        NaN   \n",
       "1451340                              NaN                        NaN   \n",
       "1451349                              NaN                        NaN   \n",
       "1451375                              NaN                        NaN   \n",
       "\n",
       "        Unnamed: 24  \n",
       "1               NaN  \n",
       "70              NaN  \n",
       "71              NaN  \n",
       "72              NaN  \n",
       "73              NaN  \n",
       "...             ...  \n",
       "1451320         NaN  \n",
       "1451331         NaN  \n",
       "1451340         NaN  \n",
       "1451349         NaN  \n",
       "1451375         NaN  \n",
       "\n",
       "[1225608 rows x 25 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snv_clinvar_df.loc[:, 'VariationID'] = snv_clinvar_df['VariationID'].apply(lambda vid: f\"ClinVar:{vid}\")\n",
    "snv_clinvar_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "408eede8-2b63-4581-a815-0bf00a434dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_alt_ref(name):\n",
    "    \"\"\"\n",
    "    Extracts the reference (ref) and alternative (alt) allele values from a given string using regular expressions.\n",
    "\n",
    "    The function works with two main types of patterns:\n",
    "    1. Genomic variants in the form of \"position ref > alt\" (e.g., \"123 A > T\").\n",
    "    2. Amino acid changes in the form of \"A123T\" (where A is the original amino acid and T is the altered amino acid).\n",
    "\n",
    "    It returns the reference and alternative alleles or amino acids depending on the input.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    name : str\n",
    "        The input string containing genomic or amino acid variant information.\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    tuple\n",
    "        A tuple containing the reference (ref) and alternative (alt) alleles or amino acids.\n",
    "        If no match is found, both `ref` and `alt` will be empty strings (\"\").\n",
    "    \"\"\"\n",
    "\n",
    "    # Regular expression pattern to match two types of variants:\n",
    "    # 1. Genomic variant: \"position ref > alt\" (e.g., \"123 A > T\").\n",
    "    # 2. Amino acid change: \"A123T\" (e.g., \"A\" for Alanine, \"T\" for Threonine).\n",
    "    pattern = r\"\\s*([\\d\\s]+)([ACGT]+)[>-]([ACGT]+)|^([A-Z])(\\d+)([A-Z])$\"\n",
    "    \n",
    "    # Search for matches in the input string using the defined pattern\n",
    "    match = re.findall(pattern, name)\n",
    "    \n",
    "    # Initialize empty strings for ref and alt\n",
    "    ref, alt = \"\", \"\"\n",
    "\n",
    "    if match:\n",
    "        # Clean up any extra spaces around the matched groups\n",
    "        match = list(map(str.strip, match[0]))\n",
    "        \n",
    "        # Separate the matching groups into two categories:\n",
    "        # 1. Genomic variant groups (position, ref, alt)\n",
    "        # 2. Amino acid change groups (original amino acid, position, altered amino acid)\n",
    "        pos_ref_alt = match[0:3]  # Groups for genomic variant (position, ref, alt)\n",
    "        aa_change = match[3:6]    # Groups for amino acid change (original AA, position, new AA)\n",
    "        \n",
    "        # Check if the input corresponds to a genomic variant (e.g., \"123 A > T\")\n",
    "        if any(pos_ref_alt):\n",
    "            ref = pos_ref_alt[1]  # Reference allele (e.g., A)\n",
    "            alt = pos_ref_alt[2]  # Alternative allele (e.g., T)\n",
    "\n",
    "        # Check if the input corresponds to an amino acid change (e.g., \"A123T\")\n",
    "        elif any(aa_change):\n",
    "            # Assuming `rev_table` is a dictionary that maps amino acid codes to full names or similar\n",
    "            ref = \",\".join(rev_table[aa_change[0]])  # Original amino acid\n",
    "            alt = \",\".join(rev_table[aa_change[2]])  # Altered amino acid\n",
    "\n",
    "    return (ref, alt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "db5c89a4-e2e1-4d2e-87df-a2c231ce9d91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1225608\n"
     ]
    }
   ],
   "source": [
    "ref_alts = [get_alt_ref(name) for name in snv_clinvar_df['Name'].values]\n",
    "print(len(ref_alts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ee393d17-de75-4169-8d84-572af0c352b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451320</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451331</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451340</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451349</th>\n",
       "      <td>GCT,GCC,GCA,GCG</td>\n",
       "      <td>AAT,AAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451375</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1225608 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ref      alt\n",
       "1                      A        G\n",
       "70                     C        T\n",
       "71                     C        T\n",
       "72                     C        T\n",
       "73                     C        T\n",
       "...                  ...      ...\n",
       "1451320                C        T\n",
       "1451331                G        A\n",
       "1451340                A        G\n",
       "1451349  GCT,GCC,GCA,GCG  AAT,AAC\n",
       "1451375                G        A\n",
       "\n",
       "[1225608 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extracted_df = pd.DataFrame(ref_alts, columns=['ref', 'alt'])\n",
    "extracted_df.index = snv_clinvar_df.index\n",
    "extracted_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "63e956f0-c890-40c4-9aa8-ad60bf7aa274",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene(s)</th>\n",
       "      <th>Protein change</th>\n",
       "      <th>Condition(s)</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GRCh37Chromosome</th>\n",
       "      <th>GRCh37Location</th>\n",
       "      <th>GRCh38Chromosome</th>\n",
       "      <th>GRCh38Location</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>...</th>\n",
       "      <th>Germline review status</th>\n",
       "      <th>Somatic clinical impact</th>\n",
       "      <th>Somatic clinical impact date last evaluated</th>\n",
       "      <th>Somatic clinical impact review status</th>\n",
       "      <th>Oncogenicity classification</th>\n",
       "      <th>Oncogenicity date last evaluated</th>\n",
       "      <th>Oncogenicity review status</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001005484.2(OR4F5):c.107A&gt;G (p.Glu36Gly)</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>E36G</td>\n",
       "      <td>not specified</td>\n",
       "      <td>VCV002205837</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>ClinVar:2205837</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.552C&gt;T (p.Ile184=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001543320</td>\n",
       "      <td>1</td>\n",
       "      <td>861336</td>\n",
       "      <td>1</td>\n",
       "      <td>925956</td>\n",
       "      <td>ClinVar:1543320</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.565C&gt;T (p.Pro189Ser)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>P10S, P189S</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001648427</td>\n",
       "      <td>1</td>\n",
       "      <td>861349</td>\n",
       "      <td>1</td>\n",
       "      <td>925969</td>\n",
       "      <td>ClinVar:1648427</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.576C&gt;T (p.Cys192=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001936488</td>\n",
       "      <td>1</td>\n",
       "      <td>861360</td>\n",
       "      <td>1</td>\n",
       "      <td>925980</td>\n",
       "      <td>ClinVar:1936488</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NM_001385641.1(SAMD11):c.582C&gt;T (p.Cys194=)</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not provided</td>\n",
       "      <td>VCV001568423</td>\n",
       "      <td>1</td>\n",
       "      <td>861366</td>\n",
       "      <td>1</td>\n",
       "      <td>925986</td>\n",
       "      <td>ClinVar:1568423</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451320</th>\n",
       "      <td>MTHFR, 1081C-T</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homocystinuria due to methylene tetrahydrofola...</td>\n",
       "      <td>VCV000003526</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:3526</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451331</th>\n",
       "      <td>NPC1, IVS16, G-A, -82</td>\n",
       "      <td>NPC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Niemann-Pick disease, type C1</td>\n",
       "      <td>VCV000002975</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:2975</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451340</th>\n",
       "      <td>NIPBL, IVS44DS, A-G, +4</td>\n",
       "      <td>NIPBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cornelia de Lange syndrome 1</td>\n",
       "      <td>VCV000002149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:2149</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451349</th>\n",
       "      <td>A114N</td>\n",
       "      <td>NIPAL4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Autosomal recessive congenital ichthyosis 6</td>\n",
       "      <td>VCV000001730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:1730</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCT,GCC,GCA,GCG</td>\n",
       "      <td>AAT,AAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451375</th>\n",
       "      <td>NF1, IVS34, G-A, +18</td>\n",
       "      <td>NF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Juvenile myelomonocytic leukemia</td>\n",
       "      <td>VCV000000351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ClinVar:351</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1225608 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Name Gene(s) Protein change  \\\n",
       "1          NM_001005484.2(OR4F5):c.107A>G (p.Glu36Gly)   OR4F5           E36G   \n",
       "70         NM_001385641.1(SAMD11):c.552C>T (p.Ile184=)  SAMD11            NaN   \n",
       "71       NM_001385641.1(SAMD11):c.565C>T (p.Pro189Ser)  SAMD11    P10S, P189S   \n",
       "72         NM_001385641.1(SAMD11):c.576C>T (p.Cys192=)  SAMD11            NaN   \n",
       "73         NM_001385641.1(SAMD11):c.582C>T (p.Cys194=)  SAMD11            NaN   \n",
       "...                                                ...     ...            ...   \n",
       "1451320                                 MTHFR, 1081C-T   MTHFR            NaN   \n",
       "1451331                          NPC1, IVS16, G-A, -82    NPC1            NaN   \n",
       "1451340                        NIPBL, IVS44DS, A-G, +4   NIPBL            NaN   \n",
       "1451349                                          A114N  NIPAL4            NaN   \n",
       "1451375                           NF1, IVS34, G-A, +18     NF1            NaN   \n",
       "\n",
       "                                              Condition(s)     Accession  \\\n",
       "1                                            not specified  VCV002205837   \n",
       "70                                            not provided  VCV001543320   \n",
       "71                                            not provided  VCV001648427   \n",
       "72                                            not provided  VCV001936488   \n",
       "73                                            not provided  VCV001568423   \n",
       "...                                                    ...           ...   \n",
       "1451320  Homocystinuria due to methylene tetrahydrofola...  VCV000003526   \n",
       "1451331                      Niemann-Pick disease, type C1  VCV000002975   \n",
       "1451340                       Cornelia de Lange syndrome 1  VCV000002149   \n",
       "1451349        Autosomal recessive congenital ichthyosis 6  VCV000001730   \n",
       "1451375                   Juvenile myelomonocytic leukemia  VCV000000351   \n",
       "\n",
       "        GRCh37Chromosome GRCh37Location GRCh38Chromosome GRCh38Location  \\\n",
       "1                      1          69134                1          69134   \n",
       "70                     1         861336                1         925956   \n",
       "71                     1         861349                1         925969   \n",
       "72                     1         861360                1         925980   \n",
       "73                     1         861366                1         925986   \n",
       "...                  ...            ...              ...            ...   \n",
       "1451320              NaN            NaN              NaN            NaN   \n",
       "1451331              NaN            NaN              NaN            NaN   \n",
       "1451340              NaN            NaN              NaN            NaN   \n",
       "1451349              NaN            NaN              NaN            NaN   \n",
       "1451375              NaN            NaN              NaN            NaN   \n",
       "\n",
       "             VariationID  ...               Germline review status  \\\n",
       "1        ClinVar:2205837  ...  criteria provided, single submitter   \n",
       "70       ClinVar:1543320  ...  criteria provided, single submitter   \n",
       "71       ClinVar:1648427  ...  criteria provided, single submitter   \n",
       "72       ClinVar:1936488  ...  criteria provided, single submitter   \n",
       "73       ClinVar:1568423  ...  criteria provided, single submitter   \n",
       "...                  ...  ...                                  ...   \n",
       "1451320     ClinVar:3526  ...       no assertion criteria provided   \n",
       "1451331     ClinVar:2975  ...       no assertion criteria provided   \n",
       "1451340     ClinVar:2149  ...       no assertion criteria provided   \n",
       "1451349     ClinVar:1730  ...       no assertion criteria provided   \n",
       "1451375      ClinVar:351  ...       no assertion criteria provided   \n",
       "\n",
       "        Somatic clinical impact Somatic clinical impact date last evaluated  \\\n",
       "1                           NaN                                         NaN   \n",
       "70                          NaN                                         NaN   \n",
       "71                          NaN                                         NaN   \n",
       "72                          NaN                                         NaN   \n",
       "73                          NaN                                         NaN   \n",
       "...                         ...                                         ...   \n",
       "1451320                     NaN                                         NaN   \n",
       "1451331                     NaN                                         NaN   \n",
       "1451340                     NaN                                         NaN   \n",
       "1451349                     NaN                                         NaN   \n",
       "1451375                     NaN                                         NaN   \n",
       "\n",
       "        Somatic clinical impact review status Oncogenicity classification  \\\n",
       "1                                         NaN                         NaN   \n",
       "70                                        NaN                         NaN   \n",
       "71                                        NaN                         NaN   \n",
       "72                                        NaN                         NaN   \n",
       "73                                        NaN                         NaN   \n",
       "...                                       ...                         ...   \n",
       "1451320                                   NaN                         NaN   \n",
       "1451331                                   NaN                         NaN   \n",
       "1451340                                   NaN                         NaN   \n",
       "1451349                                   NaN                         NaN   \n",
       "1451375                                   NaN                         NaN   \n",
       "\n",
       "        Oncogenicity date last evaluated Oncogenicity review status  \\\n",
       "1                                    NaN                        NaN   \n",
       "70                                   NaN                        NaN   \n",
       "71                                   NaN                        NaN   \n",
       "72                                   NaN                        NaN   \n",
       "73                                   NaN                        NaN   \n",
       "...                                  ...                        ...   \n",
       "1451320                              NaN                        NaN   \n",
       "1451331                              NaN                        NaN   \n",
       "1451340                              NaN                        NaN   \n",
       "1451349                              NaN                        NaN   \n",
       "1451375                              NaN                        NaN   \n",
       "\n",
       "        Unnamed: 24              ref      alt  \n",
       "1               NaN                A        G  \n",
       "70              NaN                C        T  \n",
       "71              NaN                C        T  \n",
       "72              NaN                C        T  \n",
       "73              NaN                C        T  \n",
       "...             ...              ...      ...  \n",
       "1451320         NaN                C        T  \n",
       "1451331         NaN                G        A  \n",
       "1451340         NaN                A        G  \n",
       "1451349         NaN  GCT,GCC,GCA,GCG  AAT,AAC  \n",
       "1451375         NaN                G        A  \n",
       "\n",
       "[1225608 rows x 27 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snv_clinvar_df = pd.concat([snv_clinvar_df, extracted_df], axis=1)\n",
    "snv_clinvar_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "dc55f000-ab83-4e39-8b07-488659150a57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Name', 'Gene(s)', 'Protein change', 'Condition(s)', 'Accession',\n",
       "       'GRCh37Chromosome', 'GRCh37Location', 'GRCh38Chromosome',\n",
       "       'GRCh38Location', 'VariationID', 'AlleleID(s)', 'dbSNP ID',\n",
       "       'Canonical SPDI', 'Variant type', 'Molecular consequence',\n",
       "       'Germline classification', 'Germline date last evaluated',\n",
       "       'Germline review status', 'Somatic clinical impact',\n",
       "       'Somatic clinical impact date last evaluated',\n",
       "       'Somatic clinical impact review status', 'Oncogenicity classification',\n",
       "       'Oncogenicity date last evaluated', 'Oncogenicity review status',\n",
       "       'Unnamed: 24', 'ref', 'alt'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snv_clinvar_df.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c082ab23-8e76-4b52-8d6b-8049c4c8749e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variation_id</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ClinVar:2205837</td>\n",
       "      <td>snv</td>\n",
       "      <td>Likely benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1</td>\n",
       "      <td>925956</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:1543320</td>\n",
       "      <td>snv</td>\n",
       "      <td>Likely benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1</td>\n",
       "      <td>925969</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:1648427</td>\n",
       "      <td>snv</td>\n",
       "      <td>Likely benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1</td>\n",
       "      <td>925980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:1936488</td>\n",
       "      <td>snv</td>\n",
       "      <td>Likely benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1</td>\n",
       "      <td>925986</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:1568423</td>\n",
       "      <td>snv</td>\n",
       "      <td>Likely benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451320</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:3526</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451331</th>\n",
       "      <td>NPC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ClinVar:2975</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451340</th>\n",
       "      <td>NIPBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ClinVar:2149</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451349</th>\n",
       "      <td>NIPAL4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCT,GCC,GCA,GCG</td>\n",
       "      <td>AAT,AAC</td>\n",
       "      <td>ClinVar:1730</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451375</th>\n",
       "      <td>NF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ClinVar:351</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1225608 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene  chr     pos              ref      alt     variation_id  \\\n",
       "1         OR4F5    1   69134                A        G  ClinVar:2205837   \n",
       "70       SAMD11    1  925956                C        T  ClinVar:1543320   \n",
       "71       SAMD11    1  925969                C        T  ClinVar:1648427   \n",
       "72       SAMD11    1  925980                C        T  ClinVar:1936488   \n",
       "73       SAMD11    1  925986                C        T  ClinVar:1568423   \n",
       "...         ...  ...     ...              ...      ...              ...   \n",
       "1451320   MTHFR  NaN     NaN                C        T     ClinVar:3526   \n",
       "1451331    NPC1  NaN     NaN                G        A     ClinVar:2975   \n",
       "1451340   NIPBL  NaN     NaN                A        G     ClinVar:2149   \n",
       "1451349  NIPAL4  NaN     NaN  GCT,GCC,GCA,GCG  AAT,AAC     ClinVar:1730   \n",
       "1451375     NF1  NaN     NaN                G        A      ClinVar:351   \n",
       "\n",
       "        variant_type clinical_significance  \n",
       "1                snv         Likely benign  \n",
       "70               snv         Likely benign  \n",
       "71               snv         Likely benign  \n",
       "72               snv         Likely benign  \n",
       "73               snv         Likely benign  \n",
       "...              ...                   ...  \n",
       "1451320          snv            Pathogenic  \n",
       "1451331          snv            Pathogenic  \n",
       "1451340          snv            Pathogenic  \n",
       "1451349          snv            Pathogenic  \n",
       "1451375          snv            Pathogenic  \n",
       "\n",
       "[1225608 rows x 8 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expected_col_names = ['gene', 'chr', 'pos', 'ref', 'alt', 'variation_id', 'variant_type', 'clinical_significance']\n",
    "\n",
    "clinvar_cols = ['Gene(s)','GRCh38Chromosome','GRCh38Location','ref','alt','VariationID','Variant type','Germline classification']\n",
    "snv_clinvar_df = snv_clinvar_df[clinvar_cols]\n",
    "snv_clinvar_df.columns = expected_col_names\n",
    "snv_clinvar_df.loc[:, 'variant_type'] = snv_clinvar_df['variant_type'].replace({'single nucleotide variant' : 'snv'})\n",
    "snv_clinvar_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0606c605-d673-4d92-a52b-35869f092e35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>variation</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>validation_status</th>\n",
       "      <th>function_class</th>\n",
       "      <th>gene</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>247</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.020612:103:1000Genomes|C:0.:0:ALSPAC|C:0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>snv</td>\n",
       "      <td>248</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>A:0.038257:192:1000Genomes|A:0.06876:265:ALSPA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>249</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.100718:504:1000Genomes|C:0.079917:308:ALSP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>19953512</td>\n",
       "      <td>-&gt;TG</td>\n",
       "      <td>delins</td>\n",
       "      <td>250</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>TG:0.241255:1208:1000Genomes|TG:0.123768:477:A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>251</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.057308:287:1000Genomes|C:0.000778:3:ALSPAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990134</th>\n",
       "      <td>X</td>\n",
       "      <td>46881463</td>\n",
       "      <td>-&gt;GTATTA,GTATTATTA</td>\n",
       "      <td>delins</td>\n",
       "      <td>1556328727</td>\n",
       "      <td>likely-benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>RP2</td>\n",
       "      <td>TAGTAT:0.00344:13:1000Genomes|TAGTATTATTA:0.:0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990135</th>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C&gt;A,T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556826649</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990136</th>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C&gt;A,T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556826649</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990137</th>\n",
       "      <td>X</td>\n",
       "      <td>152701711</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556827155</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>T:0.005411:20:1000Genomes|A:0.001736:151:ExAC|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990138</th>\n",
       "      <td>X</td>\n",
       "      <td>152728080</td>\n",
       "      <td>G&gt;A,C</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556830729</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>CSAG1</td>\n",
       "      <td>C:0.00728:35:1000Genomes|C:0.01666:176:GoESP|T...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>990139 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chr        pos           variation variant_type      snp_id  \\\n",
       "0         8   19953280                 A>C          snv         247   \n",
       "1         8   19953315                 G>A          snv         248   \n",
       "2         8   19953495                 T>C          snv         249   \n",
       "3         8   19953512                ->TG       delins         250   \n",
       "4         8   19953649                 T>C          snv         251   \n",
       "...     ...        ...                 ...          ...         ...   \n",
       "990134    X   46881463  ->GTATTA,GTATTATTA       delins  1556328727   \n",
       "990135    X  152701444               C>A,T          snv  1556826649   \n",
       "990136    X  152701444               C>A,T          snv  1556826649   \n",
       "990137    X  152701711                 C>T          snv  1556827155   \n",
       "990138    X  152728080               G>A,C          snv  1556830729   \n",
       "\n",
       "       clinical_significance                validation_status  \\\n",
       "0                     benign  by-frequency;by-alfa;by-cluster   \n",
       "1                     benign  by-frequency;by-alfa;by-cluster   \n",
       "2                     benign  by-frequency;by-alfa;by-cluster   \n",
       "3                     benign  by-frequency;by-alfa;by-cluster   \n",
       "4                     benign  by-frequency;by-alfa;by-cluster   \n",
       "...                      ...                              ...   \n",
       "990134         likely-benign  by-frequency;by-alfa;by-cluster   \n",
       "990135                benign  by-frequency;by-alfa;by-cluster   \n",
       "990136                benign  by-frequency;by-alfa;by-cluster   \n",
       "990137                benign  by-frequency;by-alfa;by-cluster   \n",
       "990138                benign  by-frequency;by-alfa;by-cluster   \n",
       "\n",
       "                                    function_class    gene  \\\n",
       "0                                   intron_variant     LPL   \n",
       "1       coding_sequence_variant;synonymous_variant     LPL   \n",
       "2                                   intron_variant     LPL   \n",
       "3                                   intron_variant     LPL   \n",
       "4                                   intron_variant     LPL   \n",
       "...                                            ...     ...   \n",
       "990134                         3_prime_UTR_variant     RP2   \n",
       "990135  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990136  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990137  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990138    coding_sequence_variant;missense_variant   CSAG1   \n",
       "\n",
       "                                                frequency  \n",
       "0       C:0.020612:103:1000Genomes|C:0.:0:ALSPAC|C:0.0...  \n",
       "1       A:0.038257:192:1000Genomes|A:0.06876:265:ALSPA...  \n",
       "2       C:0.100718:504:1000Genomes|C:0.079917:308:ALSP...  \n",
       "3       TG:0.241255:1208:1000Genomes|TG:0.123768:477:A...  \n",
       "4       C:0.057308:287:1000Genomes|C:0.000778:3:ALSPAC...  \n",
       "...                                                   ...  \n",
       "990134  TAGTAT:0.00344:13:1000Genomes|TAGTATTATTA:0.:0...  \n",
       "990135  A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...  \n",
       "990136  A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...  \n",
       "990137  T:0.005411:20:1000Genomes|A:0.001736:151:ExAC|...  \n",
       "990138  C:0.00728:35:1000Genomes|C:0.01666:176:GoESP|T...  \n",
       "\n",
       "[990139 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df = pd.read_csv(\"snp_result.txt\", sep=\"\\t\")\n",
    "snp_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "415a2ba2-cd16-40cf-95e1-4dc87674e3fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_type\n",
       "snv             802116\n",
       "delins          141752\n",
       "del              20611\n",
       "variant_type     19414\n",
       "ins               4549\n",
       "mnv               1697\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df['variant_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1e9e4775-4175-481a-b900-f60bb57f7d3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>variation</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>validation_status</th>\n",
       "      <th>function_class</th>\n",
       "      <th>gene</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989909</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989960</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990011</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990062</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990113</th>\n",
       "      <td>#chr</td>\n",
       "      <td>pos</td>\n",
       "      <td>variation</td>\n",
       "      <td>variant_type</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>validation_status</td>\n",
       "      <td>function_class</td>\n",
       "      <td>gene</td>\n",
       "      <td>frequency</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19414 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #chr  pos  variation  variant_type  snp_id  clinical_significance  \\\n",
       "50      #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "101     #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "152     #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "203     #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "254     #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "...      ...  ...        ...           ...     ...                    ...   \n",
       "989909  #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "989960  #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "990011  #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "990062  #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "990113  #chr  pos  variation  variant_type  snp_id  clinical_significance   \n",
       "\n",
       "        validation_status  function_class  gene  frequency  \n",
       "50      validation_status  function_class  gene  frequency  \n",
       "101     validation_status  function_class  gene  frequency  \n",
       "152     validation_status  function_class  gene  frequency  \n",
       "203     validation_status  function_class  gene  frequency  \n",
       "254     validation_status  function_class  gene  frequency  \n",
       "...                   ...             ...   ...        ...  \n",
       "989909  validation_status  function_class  gene  frequency  \n",
       "989960  validation_status  function_class  gene  frequency  \n",
       "990011  validation_status  function_class  gene  frequency  \n",
       "990062  validation_status  function_class  gene  frequency  \n",
       "990113  validation_status  function_class  gene  frequency  \n",
       "\n",
       "[19414 rows x 10 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df[snp_df['variant_type'] == 'variant_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "284536c6-0e8e-4c49-82a2-bdc1dfc844a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>variation</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>validation_status</th>\n",
       "      <th>function_class</th>\n",
       "      <th>gene</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>247</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.020612:103:1000Genomes|C:0.:0:ALSPAC|C:0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>snv</td>\n",
       "      <td>248</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>A:0.038257:192:1000Genomes|A:0.06876:265:ALSPA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>249</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.100718:504:1000Genomes|C:0.079917:308:ALSP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>19953512</td>\n",
       "      <td>-&gt;TG</td>\n",
       "      <td>delins</td>\n",
       "      <td>250</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>TG:0.241255:1208:1000Genomes|TG:0.123768:477:A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>251</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.057308:287:1000Genomes|C:0.000778:3:ALSPAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990134</th>\n",
       "      <td>X</td>\n",
       "      <td>46881463</td>\n",
       "      <td>-&gt;GTATTA,GTATTATTA</td>\n",
       "      <td>delins</td>\n",
       "      <td>1556328727</td>\n",
       "      <td>likely-benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>RP2</td>\n",
       "      <td>TAGTAT:0.00344:13:1000Genomes|TAGTATTATTA:0.:0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990135</th>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C&gt;A,T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556826649</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990136</th>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C&gt;A,T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556826649</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990137</th>\n",
       "      <td>X</td>\n",
       "      <td>152701711</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556827155</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>T:0.005411:20:1000Genomes|A:0.001736:151:ExAC|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990138</th>\n",
       "      <td>X</td>\n",
       "      <td>152728080</td>\n",
       "      <td>G&gt;A,C</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556830729</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>CSAG1</td>\n",
       "      <td>C:0.00728:35:1000Genomes|C:0.01666:176:GoESP|T...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>970725 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chr        pos           variation variant_type      snp_id  \\\n",
       "0         8   19953280                 A>C          snv         247   \n",
       "1         8   19953315                 G>A          snv         248   \n",
       "2         8   19953495                 T>C          snv         249   \n",
       "3         8   19953512                ->TG       delins         250   \n",
       "4         8   19953649                 T>C          snv         251   \n",
       "...     ...        ...                 ...          ...         ...   \n",
       "990134    X   46881463  ->GTATTA,GTATTATTA       delins  1556328727   \n",
       "990135    X  152701444               C>A,T          snv  1556826649   \n",
       "990136    X  152701444               C>A,T          snv  1556826649   \n",
       "990137    X  152701711                 C>T          snv  1556827155   \n",
       "990138    X  152728080               G>A,C          snv  1556830729   \n",
       "\n",
       "       clinical_significance                validation_status  \\\n",
       "0                     benign  by-frequency;by-alfa;by-cluster   \n",
       "1                     benign  by-frequency;by-alfa;by-cluster   \n",
       "2                     benign  by-frequency;by-alfa;by-cluster   \n",
       "3                     benign  by-frequency;by-alfa;by-cluster   \n",
       "4                     benign  by-frequency;by-alfa;by-cluster   \n",
       "...                      ...                              ...   \n",
       "990134         likely-benign  by-frequency;by-alfa;by-cluster   \n",
       "990135                benign  by-frequency;by-alfa;by-cluster   \n",
       "990136                benign  by-frequency;by-alfa;by-cluster   \n",
       "990137                benign  by-frequency;by-alfa;by-cluster   \n",
       "990138                benign  by-frequency;by-alfa;by-cluster   \n",
       "\n",
       "                                    function_class    gene  \\\n",
       "0                                   intron_variant     LPL   \n",
       "1       coding_sequence_variant;synonymous_variant     LPL   \n",
       "2                                   intron_variant     LPL   \n",
       "3                                   intron_variant     LPL   \n",
       "4                                   intron_variant     LPL   \n",
       "...                                            ...     ...   \n",
       "990134                         3_prime_UTR_variant     RP2   \n",
       "990135  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990136  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990137  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990138    coding_sequence_variant;missense_variant   CSAG1   \n",
       "\n",
       "                                                frequency  \n",
       "0       C:0.020612:103:1000Genomes|C:0.:0:ALSPAC|C:0.0...  \n",
       "1       A:0.038257:192:1000Genomes|A:0.06876:265:ALSPA...  \n",
       "2       C:0.100718:504:1000Genomes|C:0.079917:308:ALSP...  \n",
       "3       TG:0.241255:1208:1000Genomes|TG:0.123768:477:A...  \n",
       "4       C:0.057308:287:1000Genomes|C:0.000778:3:ALSPAC...  \n",
       "...                                                   ...  \n",
       "990134  TAGTAT:0.00344:13:1000Genomes|TAGTATTATTA:0.:0...  \n",
       "990135  A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...  \n",
       "990136  A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...  \n",
       "990137  T:0.005411:20:1000Genomes|A:0.001736:151:ExAC|...  \n",
       "990138  C:0.00728:35:1000Genomes|C:0.01666:176:GoESP|T...  \n",
       "\n",
       "[970725 rows x 10 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df = snp_df[snp_df['variant_type'] != 'variant_type']\n",
    "snp_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f6db4cff-1121-45db-b59a-8c8b012a04e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>variation</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>validation_status</th>\n",
       "      <th>function_class</th>\n",
       "      <th>gene</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>990119</th>\n",
       "      <td>20</td>\n",
       "      <td>2652733</td>\n",
       "      <td>GGCCTGGGCCTG&gt;-,GGCCTG,GGCCTGGGCCTGGGCCTG,GGCCT...</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555779353</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant;upstream_transcript_variant;2KB...</td>\n",
       "      <td>NOP56;SNORD110;MIR1292;LOC105372505</td>\n",
       "      <td>GGGCCTGGGCCTGGGCCTG:0.:0:ALFA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990120</th>\n",
       "      <td>20</td>\n",
       "      <td>2652733</td>\n",
       "      <td>GGCCTGGGCCTG&gt;-,GGCCTG,GGCCTGGGCCTGGGCCTG,GGCCT...</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555779353</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant;upstream_transcript_variant;2KB...</td>\n",
       "      <td>NOP56;SNORD110;MIR1292;LOC105372505</td>\n",
       "      <td>GGGCCTGGGCCTGGGCCTG:0.:0:ALFA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990121</th>\n",
       "      <td>19</td>\n",
       "      <td>45409896</td>\n",
       "      <td>-&gt;AT,ATCT,ATTAT,ATTATTAT,ATTATTATTAT</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555785259</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_downstream_transcript_variant;intron_var...</td>\n",
       "      <td>ERCC1;POLR1G</td>\n",
       "      <td>TATTA:0.03094:43:Korea1K|TATC:0.00575:3:Northe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990122</th>\n",
       "      <td>20</td>\n",
       "      <td>3916912</td>\n",
       "      <td>-&gt;CCCCCT,CCCCT</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555789541</td>\n",
       "      <td>likely-benign;benign-likely-benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_downstream_transcript_variant;intron_var...</td>\n",
       "      <td>PANK2;MIR103A2</td>\n",
       "      <td>TCCCCC:0.018613:233:GoESP|TCCCC:0.000546:1:Kor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990123</th>\n",
       "      <td>20</td>\n",
       "      <td>10644276</td>\n",
       "      <td>-&gt;TCAC,TCACAC,TCACACAAACAC,TCACACAC,TCACACACAC...</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555828066</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>JAG1</td>\n",
       "      <td>CTCACA:0.3396:1309:ALSPAC|CTCACA:0.3433:1273:T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990124</th>\n",
       "      <td>X</td>\n",
       "      <td>2856800</td>\n",
       "      <td>TATC&gt;-,TATCTATC,TATCTATCTATC</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555898880</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>GYG2</td>\n",
       "      <td>ATCTATCT:0.00002:2:GnomAD|-:0.36977:4739:TOMMO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990125</th>\n",
       "      <td>22</td>\n",
       "      <td>40345589</td>\n",
       "      <td>GGG&gt;-,G,GG,GGGG,GGGGG,GGGGGG,GGGGGGG</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555903152</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>upstream_transcript_variant;2KB_upstream_variant</td>\n",
       "      <td>ADSL</td>\n",
       "      <td>-:0.001645:3:Korea1K|-:0.000015:4:TOPMED|GGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990126</th>\n",
       "      <td>22</td>\n",
       "      <td>40345589</td>\n",
       "      <td>GGG&gt;-,G,GG,GGGG,GGGGG,GGGGGG,GGGGGGG</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555903152</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>upstream_transcript_variant;2KB_upstream_variant</td>\n",
       "      <td>ADSL</td>\n",
       "      <td>-:0.001645:3:Korea1K|-:0.000015:4:TOPMED|GGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990127</th>\n",
       "      <td>22</td>\n",
       "      <td>40345589</td>\n",
       "      <td>GGG&gt;-,G,GG,GGGG,GGGGG,GGGGGG,GGGGGGG</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555903152</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>upstream_transcript_variant;2KB_upstream_variant</td>\n",
       "      <td>ADSL</td>\n",
       "      <td>-:0.001645:3:Korea1K|-:0.000015:4:TOPMED|GGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990128</th>\n",
       "      <td>22</td>\n",
       "      <td>33432596</td>\n",
       "      <td>GCAT&gt;-,GCATGCAT,GCATGCATGCAT,GCATGCATGCATGCAT</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555916106</td>\n",
       "      <td>likely-benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_upstream_transcript_variant;intron_variant</td>\n",
       "      <td>LARGE1</td>\n",
       "      <td>CATG:0.0052:87:TOMMO|CATGCATGCATGCATGCATGCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990129</th>\n",
       "      <td>21</td>\n",
       "      <td>46353752</td>\n",
       "      <td>GAGAGA&gt;-,GA,GAGA,GAGAGAGA,GAGAGAGAGA,GAGAGAGAGAGA</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555954604</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_upstream_transcript_variant;intron_variant</td>\n",
       "      <td>PCNT</td>\n",
       "      <td>GA:0.02024:37:Korea1K|-:0.03667:22:NorthernSwe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990130</th>\n",
       "      <td>21</td>\n",
       "      <td>46353752</td>\n",
       "      <td>GAGAGA&gt;-,GA,GAGA,GAGAGAGA,GAGAGAGAGA,GAGAGAGAGAGA</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555954604</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_upstream_transcript_variant;intron_variant</td>\n",
       "      <td>PCNT</td>\n",
       "      <td>GA:0.02024:37:Korea1K|-:0.03667:22:NorthernSwe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990131</th>\n",
       "      <td>21</td>\n",
       "      <td>46353752</td>\n",
       "      <td>GAGAGA&gt;-,GA,GAGA,GAGAGAGA,GAGAGAGAGA,GAGAGAGAGAGA</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555954604</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_upstream_transcript_variant;intron_variant</td>\n",
       "      <td>PCNT</td>\n",
       "      <td>GA:0.02024:37:Korea1K|-:0.03667:22:NorthernSwe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990132</th>\n",
       "      <td>21</td>\n",
       "      <td>46353752</td>\n",
       "      <td>GAGAGA&gt;-,GA,GAGA,GAGAGAGA,GAGAGAGAGA,GAGAGAGAGAGA</td>\n",
       "      <td>delins</td>\n",
       "      <td>1555954604</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_upstream_transcript_variant;intron_variant</td>\n",
       "      <td>PCNT</td>\n",
       "      <td>GA:0.02024:37:Korea1K|-:0.03667:22:NorthernSwe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990133</th>\n",
       "      <td>X</td>\n",
       "      <td>120429176</td>\n",
       "      <td>CACA&gt;-,CA,CACACA</td>\n",
       "      <td>delins</td>\n",
       "      <td>1556075606</td>\n",
       "      <td>likely-benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_downstream_transcript_variant;3_prime_UT...</td>\n",
       "      <td>LAMP2</td>\n",
       "      <td>AC:0.003179:12:1000Genomes|AC:0.000045:1:TOMMO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990134</th>\n",
       "      <td>X</td>\n",
       "      <td>46881463</td>\n",
       "      <td>-&gt;GTATTA,GTATTATTA</td>\n",
       "      <td>delins</td>\n",
       "      <td>1556328727</td>\n",
       "      <td>likely-benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>RP2</td>\n",
       "      <td>TAGTAT:0.00344:13:1000Genomes|TAGTATTATTA:0.:0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990135</th>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C&gt;A,T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556826649</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990136</th>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C&gt;A,T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556826649</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990137</th>\n",
       "      <td>X</td>\n",
       "      <td>152701711</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556827155</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>T:0.005411:20:1000Genomes|A:0.001736:151:ExAC|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990138</th>\n",
       "      <td>X</td>\n",
       "      <td>152728080</td>\n",
       "      <td>G&gt;A,C</td>\n",
       "      <td>snv</td>\n",
       "      <td>1556830729</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>CSAG1</td>\n",
       "      <td>C:0.00728:35:1000Genomes|C:0.01666:176:GoESP|T...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chr        pos                                          variation  \\\n",
       "990119   20    2652733  GGCCTGGGCCTG>-,GGCCTG,GGCCTGGGCCTGGGCCTG,GGCCT...   \n",
       "990120   20    2652733  GGCCTGGGCCTG>-,GGCCTG,GGCCTGGGCCTGGGCCTG,GGCCT...   \n",
       "990121   19   45409896               ->AT,ATCT,ATTAT,ATTATTAT,ATTATTATTAT   \n",
       "990122   20    3916912                                     ->CCCCCT,CCCCT   \n",
       "990123   20   10644276  ->TCAC,TCACAC,TCACACAAACAC,TCACACAC,TCACACACAC...   \n",
       "990124    X    2856800                       TATC>-,TATCTATC,TATCTATCTATC   \n",
       "990125   22   40345589               GGG>-,G,GG,GGGG,GGGGG,GGGGGG,GGGGGGG   \n",
       "990126   22   40345589               GGG>-,G,GG,GGGG,GGGGG,GGGGGG,GGGGGGG   \n",
       "990127   22   40345589               GGG>-,G,GG,GGGG,GGGGG,GGGGGG,GGGGGGG   \n",
       "990128   22   33432596      GCAT>-,GCATGCAT,GCATGCATGCAT,GCATGCATGCATGCAT   \n",
       "990129   21   46353752  GAGAGA>-,GA,GAGA,GAGAGAGA,GAGAGAGAGA,GAGAGAGAGAGA   \n",
       "990130   21   46353752  GAGAGA>-,GA,GAGA,GAGAGAGA,GAGAGAGAGA,GAGAGAGAGAGA   \n",
       "990131   21   46353752  GAGAGA>-,GA,GAGA,GAGAGAGA,GAGAGAGAGA,GAGAGAGAGAGA   \n",
       "990132   21   46353752  GAGAGA>-,GA,GAGA,GAGAGAGA,GAGAGAGAGA,GAGAGAGAGAGA   \n",
       "990133    X  120429176                                   CACA>-,CA,CACACA   \n",
       "990134    X   46881463                                 ->GTATTA,GTATTATTA   \n",
       "990135    X  152701444                                              C>A,T   \n",
       "990136    X  152701444                                              C>A,T   \n",
       "990137    X  152701711                                                C>T   \n",
       "990138    X  152728080                                              G>A,C   \n",
       "\n",
       "       variant_type      snp_id               clinical_significance  \\\n",
       "990119       delins  1555779353                          pathogenic   \n",
       "990120       delins  1555779353                          pathogenic   \n",
       "990121       delins  1555785259                              benign   \n",
       "990122       delins  1555789541  likely-benign;benign-likely-benign   \n",
       "990123       delins  1555828066                              benign   \n",
       "990124       delins  1555898880                              benign   \n",
       "990125       delins  1555903152                              benign   \n",
       "990126       delins  1555903152                              benign   \n",
       "990127       delins  1555903152                              benign   \n",
       "990128       delins  1555916106                       likely-benign   \n",
       "990129       delins  1555954604                              benign   \n",
       "990130       delins  1555954604                              benign   \n",
       "990131       delins  1555954604                              benign   \n",
       "990132       delins  1555954604                              benign   \n",
       "990133       delins  1556075606                       likely-benign   \n",
       "990134       delins  1556328727                       likely-benign   \n",
       "990135          snv  1556826649                              benign   \n",
       "990136          snv  1556826649                              benign   \n",
       "990137          snv  1556827155                              benign   \n",
       "990138          snv  1556830729                              benign   \n",
       "\n",
       "                      validation_status  \\\n",
       "990119  by-frequency;by-alfa;by-cluster   \n",
       "990120  by-frequency;by-alfa;by-cluster   \n",
       "990121  by-frequency;by-alfa;by-cluster   \n",
       "990122  by-frequency;by-alfa;by-cluster   \n",
       "990123  by-frequency;by-alfa;by-cluster   \n",
       "990124  by-frequency;by-alfa;by-cluster   \n",
       "990125  by-frequency;by-alfa;by-cluster   \n",
       "990126  by-frequency;by-alfa;by-cluster   \n",
       "990127  by-frequency;by-alfa;by-cluster   \n",
       "990128  by-frequency;by-alfa;by-cluster   \n",
       "990129  by-frequency;by-alfa;by-cluster   \n",
       "990130  by-frequency;by-alfa;by-cluster   \n",
       "990131  by-frequency;by-alfa;by-cluster   \n",
       "990132  by-frequency;by-alfa;by-cluster   \n",
       "990133  by-frequency;by-alfa;by-cluster   \n",
       "990134  by-frequency;by-alfa;by-cluster   \n",
       "990135  by-frequency;by-alfa;by-cluster   \n",
       "990136  by-frequency;by-alfa;by-cluster   \n",
       "990137  by-frequency;by-alfa;by-cluster   \n",
       "990138  by-frequency;by-alfa;by-cluster   \n",
       "\n",
       "                                           function_class  \\\n",
       "990119  intron_variant;upstream_transcript_variant;2KB...   \n",
       "990120  intron_variant;upstream_transcript_variant;2KB...   \n",
       "990121  genic_downstream_transcript_variant;intron_var...   \n",
       "990122  genic_downstream_transcript_variant;intron_var...   \n",
       "990123                                     intron_variant   \n",
       "990124                                     intron_variant   \n",
       "990125   upstream_transcript_variant;2KB_upstream_variant   \n",
       "990126   upstream_transcript_variant;2KB_upstream_variant   \n",
       "990127   upstream_transcript_variant;2KB_upstream_variant   \n",
       "990128   genic_upstream_transcript_variant;intron_variant   \n",
       "990129   genic_upstream_transcript_variant;intron_variant   \n",
       "990130   genic_upstream_transcript_variant;intron_variant   \n",
       "990131   genic_upstream_transcript_variant;intron_variant   \n",
       "990132   genic_upstream_transcript_variant;intron_variant   \n",
       "990133  genic_downstream_transcript_variant;3_prime_UT...   \n",
       "990134                                3_prime_UTR_variant   \n",
       "990135         coding_sequence_variant;synonymous_variant   \n",
       "990136         coding_sequence_variant;synonymous_variant   \n",
       "990137         coding_sequence_variant;synonymous_variant   \n",
       "990138           coding_sequence_variant;missense_variant   \n",
       "\n",
       "                                       gene  \\\n",
       "990119  NOP56;SNORD110;MIR1292;LOC105372505   \n",
       "990120  NOP56;SNORD110;MIR1292;LOC105372505   \n",
       "990121                         ERCC1;POLR1G   \n",
       "990122                       PANK2;MIR103A2   \n",
       "990123                                 JAG1   \n",
       "990124                                 GYG2   \n",
       "990125                                 ADSL   \n",
       "990126                                 ADSL   \n",
       "990127                                 ADSL   \n",
       "990128                               LARGE1   \n",
       "990129                                 PCNT   \n",
       "990130                                 PCNT   \n",
       "990131                                 PCNT   \n",
       "990132                                 PCNT   \n",
       "990133                                LAMP2   \n",
       "990134                                  RP2   \n",
       "990135                               MAGEA3   \n",
       "990136                               MAGEA3   \n",
       "990137                               MAGEA3   \n",
       "990138                                CSAG1   \n",
       "\n",
       "                                                frequency  \n",
       "990119                      GGGCCTGGGCCTGGGCCTG:0.:0:ALFA  \n",
       "990120                      GGGCCTGGGCCTGGGCCTG:0.:0:ALFA  \n",
       "990121  TATTA:0.03094:43:Korea1K|TATC:0.00575:3:Northe...  \n",
       "990122  TCCCCC:0.018613:233:GoESP|TCCCC:0.000546:1:Kor...  \n",
       "990123  CTCACA:0.3396:1309:ALSPAC|CTCACA:0.3433:1273:T...  \n",
       "990124  ATCTATCT:0.00002:2:GnomAD|-:0.36977:4739:TOMMO...  \n",
       "990125  -:0.001645:3:Korea1K|-:0.000015:4:TOPMED|GGGGG...  \n",
       "990126  -:0.001645:3:Korea1K|-:0.000015:4:TOPMED|GGGGG...  \n",
       "990127  -:0.001645:3:Korea1K|-:0.000015:4:TOPMED|GGGGG...  \n",
       "990128  CATG:0.0052:87:TOMMO|CATGCATGCATGCATGCATGCATGC...  \n",
       "990129  GA:0.02024:37:Korea1K|-:0.03667:22:NorthernSwe...  \n",
       "990130  GA:0.02024:37:Korea1K|-:0.03667:22:NorthernSwe...  \n",
       "990131  GA:0.02024:37:Korea1K|-:0.03667:22:NorthernSwe...  \n",
       "990132  GA:0.02024:37:Korea1K|-:0.03667:22:NorthernSwe...  \n",
       "990133  AC:0.003179:12:1000Genomes|AC:0.000045:1:TOMMO...  \n",
       "990134  TAGTAT:0.00344:13:1000Genomes|TAGTATTATTA:0.:0...  \n",
       "990135  A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...  \n",
       "990136  A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...  \n",
       "990137  T:0.005411:20:1000Genomes|A:0.001736:151:ExAC|...  \n",
       "990138  C:0.00728:35:1000Genomes|C:0.01666:176:GoESP|T...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df.tail(20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8e610798-660b-4886-b5b1-62578a0e5b8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "970725"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_ref_alts = [variant.strip().split(\">\") for variant in snp_df['variation']]\n",
    "len(snp_ref_alts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ed168233-3d5d-4a57-8cc5-911073e41f90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-</td>\n",
       "      <td>TG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990134</th>\n",
       "      <td>-</td>\n",
       "      <td>GTATTA,GTATTATTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990135</th>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990136</th>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990137</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990138</th>\n",
       "      <td>G</td>\n",
       "      <td>A,C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>970725 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ref               alt\n",
       "0        A                 C\n",
       "1        G                 A\n",
       "2        T                 C\n",
       "3        -                TG\n",
       "4        T                 C\n",
       "...     ..               ...\n",
       "990134   -  GTATTA,GTATTATTA\n",
       "990135   C               A,T\n",
       "990136   C               A,T\n",
       "990137   C                 T\n",
       "990138   G               A,C\n",
       "\n",
       "[970725 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extracted_snp = pd.DataFrame(snp_ref_alts, columns = ['ref', 'alt'])\n",
    "extracted_snp.index = snp_df.index\n",
    "extracted_snp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7ef1401c-a38e-4cbd-860c-d4f2f0cd655f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>variation</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>validation_status</th>\n",
       "      <th>function_class</th>\n",
       "      <th>gene</th>\n",
       "      <th>frequency</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>SNP:247</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.020612:103:1000Genomes|C:0.:0:ALSPAC|C:0.0...</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>snv</td>\n",
       "      <td>SNP:248</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>A:0.038257:192:1000Genomes|A:0.06876:265:ALSPA...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>SNP:249</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.100718:504:1000Genomes|C:0.079917:308:ALSP...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>19953512</td>\n",
       "      <td>-&gt;TG</td>\n",
       "      <td>delins</td>\n",
       "      <td>SNP:250</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>TG:0.241255:1208:1000Genomes|TG:0.123768:477:A...</td>\n",
       "      <td>-</td>\n",
       "      <td>TG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>snv</td>\n",
       "      <td>SNP:251</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LPL</td>\n",
       "      <td>C:0.057308:287:1000Genomes|C:0.000778:3:ALSPAC...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990134</th>\n",
       "      <td>X</td>\n",
       "      <td>46881463</td>\n",
       "      <td>-&gt;GTATTA,GTATTATTA</td>\n",
       "      <td>delins</td>\n",
       "      <td>SNP:1556328727</td>\n",
       "      <td>likely-benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>RP2</td>\n",
       "      <td>TAGTAT:0.00344:13:1000Genomes|TAGTATTATTA:0.:0...</td>\n",
       "      <td>-</td>\n",
       "      <td>GTATTA,GTATTATTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990135</th>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C&gt;A,T</td>\n",
       "      <td>snv</td>\n",
       "      <td>SNP:1556826649</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990136</th>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C&gt;A,T</td>\n",
       "      <td>snv</td>\n",
       "      <td>SNP:1556826649</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990137</th>\n",
       "      <td>X</td>\n",
       "      <td>152701711</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>snv</td>\n",
       "      <td>SNP:1556827155</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;synonymous_variant</td>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>T:0.005411:20:1000Genomes|A:0.001736:151:ExAC|...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990138</th>\n",
       "      <td>X</td>\n",
       "      <td>152728080</td>\n",
       "      <td>G&gt;A,C</td>\n",
       "      <td>snv</td>\n",
       "      <td>SNP:1556830729</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>CSAG1</td>\n",
       "      <td>C:0.00728:35:1000Genomes|C:0.01666:176:GoESP|T...</td>\n",
       "      <td>G</td>\n",
       "      <td>A,C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>970725 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chr        pos           variation variant_type          snp_id  \\\n",
       "0         8   19953280                 A>C          snv         SNP:247   \n",
       "1         8   19953315                 G>A          snv         SNP:248   \n",
       "2         8   19953495                 T>C          snv         SNP:249   \n",
       "3         8   19953512                ->TG       delins         SNP:250   \n",
       "4         8   19953649                 T>C          snv         SNP:251   \n",
       "...     ...        ...                 ...          ...             ...   \n",
       "990134    X   46881463  ->GTATTA,GTATTATTA       delins  SNP:1556328727   \n",
       "990135    X  152701444               C>A,T          snv  SNP:1556826649   \n",
       "990136    X  152701444               C>A,T          snv  SNP:1556826649   \n",
       "990137    X  152701711                 C>T          snv  SNP:1556827155   \n",
       "990138    X  152728080               G>A,C          snv  SNP:1556830729   \n",
       "\n",
       "       clinical_significance                validation_status  \\\n",
       "0                     benign  by-frequency;by-alfa;by-cluster   \n",
       "1                     benign  by-frequency;by-alfa;by-cluster   \n",
       "2                     benign  by-frequency;by-alfa;by-cluster   \n",
       "3                     benign  by-frequency;by-alfa;by-cluster   \n",
       "4                     benign  by-frequency;by-alfa;by-cluster   \n",
       "...                      ...                              ...   \n",
       "990134         likely-benign  by-frequency;by-alfa;by-cluster   \n",
       "990135                benign  by-frequency;by-alfa;by-cluster   \n",
       "990136                benign  by-frequency;by-alfa;by-cluster   \n",
       "990137                benign  by-frequency;by-alfa;by-cluster   \n",
       "990138                benign  by-frequency;by-alfa;by-cluster   \n",
       "\n",
       "                                    function_class    gene  \\\n",
       "0                                   intron_variant     LPL   \n",
       "1       coding_sequence_variant;synonymous_variant     LPL   \n",
       "2                                   intron_variant     LPL   \n",
       "3                                   intron_variant     LPL   \n",
       "4                                   intron_variant     LPL   \n",
       "...                                            ...     ...   \n",
       "990134                         3_prime_UTR_variant     RP2   \n",
       "990135  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990136  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990137  coding_sequence_variant;synonymous_variant  MAGEA3   \n",
       "990138    coding_sequence_variant;missense_variant   CSAG1   \n",
       "\n",
       "                                                frequency ref  \\\n",
       "0       C:0.020612:103:1000Genomes|C:0.:0:ALSPAC|C:0.0...   A   \n",
       "1       A:0.038257:192:1000Genomes|A:0.06876:265:ALSPA...   G   \n",
       "2       C:0.100718:504:1000Genomes|C:0.079917:308:ALSP...   T   \n",
       "3       TG:0.241255:1208:1000Genomes|TG:0.123768:477:A...   -   \n",
       "4       C:0.057308:287:1000Genomes|C:0.000778:3:ALSPAC...   T   \n",
       "...                                                   ...  ..   \n",
       "990134  TAGTAT:0.00344:13:1000Genomes|TAGTATTATTA:0.:0...   -   \n",
       "990135  A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...   C   \n",
       "990136  A:0.02702:102:1000Genomes|T:0.026419:2747:Gnom...   C   \n",
       "990137  T:0.005411:20:1000Genomes|A:0.001736:151:ExAC|...   C   \n",
       "990138  C:0.00728:35:1000Genomes|C:0.01666:176:GoESP|T...   G   \n",
       "\n",
       "                     alt  \n",
       "0                      C  \n",
       "1                      A  \n",
       "2                      C  \n",
       "3                     TG  \n",
       "4                      C  \n",
       "...                  ...  \n",
       "990134  GTATTA,GTATTATTA  \n",
       "990135               A,T  \n",
       "990136               A,T  \n",
       "990137                 T  \n",
       "990138               A,C  \n",
       "\n",
       "[970725 rows x 12 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df.loc[:, 'snp_id'] = snp_df['snp_id'].apply(lambda sid: f\"SNP:{sid}\")\n",
    "snp_df = pd.concat([snp_df, extracted_snp], axis=1)\n",
    "snp_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2dec04cf-4c74-4eb1-9fc5-b560714afd1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#chr', 'pos', 'variation', 'variant_type', 'snp_id',\n",
       "       'clinical_significance', 'validation_status', 'function_class', 'gene',\n",
       "       'frequency', 'ref', 'alt'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5af4bace-2c1f-405b-a279-abdfad1863b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variation_id</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:247</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP:248</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:249</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953512</td>\n",
       "      <td>-</td>\n",
       "      <td>TG</td>\n",
       "      <td>SNP:250</td>\n",
       "      <td>delins</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:251</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990134</th>\n",
       "      <td>RP2</td>\n",
       "      <td>X</td>\n",
       "      <td>46881463</td>\n",
       "      <td>-</td>\n",
       "      <td>GTATTA,GTATTATTA</td>\n",
       "      <td>SNP:1556328727</td>\n",
       "      <td>delins</td>\n",
       "      <td>likely-benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990135</th>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNP:1556826649</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990136</th>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>X</td>\n",
       "      <td>152701444</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNP:1556826649</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990137</th>\n",
       "      <td>MAGEA3</td>\n",
       "      <td>X</td>\n",
       "      <td>152701711</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP:1556827155</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990138</th>\n",
       "      <td>CSAG1</td>\n",
       "      <td>X</td>\n",
       "      <td>152728080</td>\n",
       "      <td>G</td>\n",
       "      <td>A,C</td>\n",
       "      <td>SNP:1556830729</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>970725 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene chr        pos ref               alt    variation_id  \\\n",
       "0          LPL   8   19953280   A                 C         SNP:247   \n",
       "1          LPL   8   19953315   G                 A         SNP:248   \n",
       "2          LPL   8   19953495   T                 C         SNP:249   \n",
       "3          LPL   8   19953512   -                TG         SNP:250   \n",
       "4          LPL   8   19953649   T                 C         SNP:251   \n",
       "...        ...  ..        ...  ..               ...             ...   \n",
       "990134     RP2   X   46881463   -  GTATTA,GTATTATTA  SNP:1556328727   \n",
       "990135  MAGEA3   X  152701444   C               A,T  SNP:1556826649   \n",
       "990136  MAGEA3   X  152701444   C               A,T  SNP:1556826649   \n",
       "990137  MAGEA3   X  152701711   C                 T  SNP:1556827155   \n",
       "990138   CSAG1   X  152728080   G               A,C  SNP:1556830729   \n",
       "\n",
       "       variant_type clinical_significance  \n",
       "0               snv                benign  \n",
       "1               snv                benign  \n",
       "2               snv                benign  \n",
       "3            delins                benign  \n",
       "4               snv                benign  \n",
       "...             ...                   ...  \n",
       "990134       delins         likely-benign  \n",
       "990135          snv                benign  \n",
       "990136          snv                benign  \n",
       "990137          snv                benign  \n",
       "990138          snv                benign  \n",
       "\n",
       "[970725 rows x 8 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_columns = ['gene', '#chr', 'pos', 'ref', 'alt', 'snp_id', 'variant_type', 'clinical_significance']\n",
    "snp_df = snp_df[snp_columns]\n",
    "snp_df.columns = expected_col_names\n",
    "snp_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6f944b3a-a297-4007-9f0f-ee2b44a69bee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variation_id</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:247</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP:248</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:249</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953512</td>\n",
       "      <td>-</td>\n",
       "      <td>TG</td>\n",
       "      <td>SNP:250</td>\n",
       "      <td>delins</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:251</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451320</th>\n",
       "      <td>MTHFR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:3526</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451331</th>\n",
       "      <td>NPC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ClinVar:2975</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451340</th>\n",
       "      <td>NIPBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ClinVar:2149</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451349</th>\n",
       "      <td>NIPAL4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCT,GCC,GCA,GCG</td>\n",
       "      <td>AAT,AAC</td>\n",
       "      <td>ClinVar:1730</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451375</th>\n",
       "      <td>NF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ClinVar:351</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2196333 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene  chr       pos              ref      alt  variation_id  \\\n",
       "0           LPL    8  19953280                A        C       SNP:247   \n",
       "1           LPL    8  19953315                G        A       SNP:248   \n",
       "2           LPL    8  19953495                T        C       SNP:249   \n",
       "3           LPL    8  19953512                -       TG       SNP:250   \n",
       "4           LPL    8  19953649                T        C       SNP:251   \n",
       "...         ...  ...       ...              ...      ...           ...   \n",
       "1451320   MTHFR  NaN       NaN                C        T  ClinVar:3526   \n",
       "1451331    NPC1  NaN       NaN                G        A  ClinVar:2975   \n",
       "1451340   NIPBL  NaN       NaN                A        G  ClinVar:2149   \n",
       "1451349  NIPAL4  NaN       NaN  GCT,GCC,GCA,GCG  AAT,AAC  ClinVar:1730   \n",
       "1451375     NF1  NaN       NaN                G        A   ClinVar:351   \n",
       "\n",
       "        variant_type clinical_significance  \n",
       "0                snv                benign  \n",
       "1                snv                benign  \n",
       "2                snv                benign  \n",
       "3             delins                benign  \n",
       "4                snv                benign  \n",
       "...              ...                   ...  \n",
       "1451320          snv            Pathogenic  \n",
       "1451331          snv            Pathogenic  \n",
       "1451340          snv            Pathogenic  \n",
       "1451349          snv            Pathogenic  \n",
       "1451375          snv            Pathogenic  \n",
       "\n",
       "[2196333 rows x 8 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df = pd.concat([snp_df, snv_clinvar_df], axis=0)\n",
    "final_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "745d50e1-d09d-479e-a6b1-8732510c0577",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene                     615\n",
       "chr                      399\n",
       "pos                      399\n",
       "ref                        0\n",
       "alt                        0\n",
       "variation_id               0\n",
       "variant_type               0\n",
       "clinical_significance      0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df.isna().sum()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ca0d330e-658b-4acf-9e14-1dc00487975f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variation_id</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:247</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP:248</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:249</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953512</td>\n",
       "      <td>-</td>\n",
       "      <td>TG</td>\n",
       "      <td>SNP:250</td>\n",
       "      <td>delins</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:251</td>\n",
       "      <td>snv</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448523</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15990</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:9570</td>\n",
       "      <td>snv</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448524</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15992</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:690268</td>\n",
       "      <td>snv</td>\n",
       "      <td>Likely benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448525</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16002</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ClinVar:690271</td>\n",
       "      <td>snv</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448526</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16011</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ClinVar:690273</td>\n",
       "      <td>snv</td>\n",
       "      <td>Likely benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448527</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16017</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ClinVar:690275</td>\n",
       "      <td>snv</td>\n",
       "      <td>Benign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2195340 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene chr       pos ref alt    variation_id variant_type  \\\n",
       "0          LPL   8  19953280   A   C         SNP:247          snv   \n",
       "1          LPL   8  19953315   G   A         SNP:248          snv   \n",
       "2          LPL   8  19953495   T   C         SNP:249          snv   \n",
       "3          LPL   8  19953512   -  TG         SNP:250       delins   \n",
       "4          LPL   8  19953649   T   C         SNP:251          snv   \n",
       "...        ...  ..       ...  ..  ..             ...          ...   \n",
       "1448523  MT-TP  MT     15990   C   T    ClinVar:9570          snv   \n",
       "1448524  MT-TP  MT     15992   A   T  ClinVar:690268          snv   \n",
       "1448525  MT-TP  MT     16002   T   C  ClinVar:690271          snv   \n",
       "1448526  MT-TP  MT     16011   A   G  ClinVar:690273          snv   \n",
       "1448527  MT-TP  MT     16017   T   C  ClinVar:690275          snv   \n",
       "\n",
       "                clinical_significance  \n",
       "0                              benign  \n",
       "1                              benign  \n",
       "2                              benign  \n",
       "3                              benign  \n",
       "4                              benign  \n",
       "...                               ...  \n",
       "1448523             Likely pathogenic  \n",
       "1448524                 Likely benign  \n",
       "1448525  Pathogenic/Likely pathogenic  \n",
       "1448526                 Likely benign  \n",
       "1448527                        Benign  \n",
       "\n",
       "[2195340 rows x 8 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df = final_df.dropna()\n",
    "final_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0083221f-58bc-49b7-8167-4f70432846eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variation_id</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:247</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP:248</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:249</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953512</td>\n",
       "      <td>-</td>\n",
       "      <td>TG</td>\n",
       "      <td>SNP:250</td>\n",
       "      <td>delins</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:251</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448523</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15990</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:9570</td>\n",
       "      <td>snv</td>\n",
       "      <td>harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448524</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15992</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:690268</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448525</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16002</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ClinVar:690271</td>\n",
       "      <td>snv</td>\n",
       "      <td>harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448526</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16011</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ClinVar:690273</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448527</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16017</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ClinVar:690275</td>\n",
       "      <td>snv</td>\n",
       "      <td>harmful</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2195340 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene chr       pos ref alt    variation_id variant_type  \\\n",
       "0          LPL   8  19953280   A   C         SNP:247          snv   \n",
       "1          LPL   8  19953315   G   A         SNP:248          snv   \n",
       "2          LPL   8  19953495   T   C         SNP:249          snv   \n",
       "3          LPL   8  19953512   -  TG         SNP:250       delins   \n",
       "4          LPL   8  19953649   T   C         SNP:251          snv   \n",
       "...        ...  ..       ...  ..  ..             ...          ...   \n",
       "1448523  MT-TP  MT     15990   C   T    ClinVar:9570          snv   \n",
       "1448524  MT-TP  MT     15992   A   T  ClinVar:690268          snv   \n",
       "1448525  MT-TP  MT     16002   T   C  ClinVar:690271          snv   \n",
       "1448526  MT-TP  MT     16011   A   G  ClinVar:690273          snv   \n",
       "1448527  MT-TP  MT     16017   T   C  ClinVar:690275          snv   \n",
       "\n",
       "        clinical_significance  \n",
       "0                 non-harmful  \n",
       "1                 non-harmful  \n",
       "2                 non-harmful  \n",
       "3                 non-harmful  \n",
       "4                 non-harmful  \n",
       "...                       ...  \n",
       "1448523               harmful  \n",
       "1448524           non-harmful  \n",
       "1448525               harmful  \n",
       "1448526           non-harmful  \n",
       "1448527               harmful  \n",
       "\n",
       "[2195340 rows x 8 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df.loc[:, 'clinical_significance'] = final_df['clinical_significance'].apply(lambda sig: \"non-harmful\" if \"benign\" in sig else \"harmful\")\n",
    "final_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "99b5701d-9792-4dd4-8895-9d04f28b7e71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([1 for var in final_df['ref'] if len(var.split(\",\")) > 1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8fb1003e-cdc1-4a14-9a5b-fde751d8f436",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "387444"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([1 for var in final_df['alt'] if len(var.split(\",\")) > 1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "58ea5c0a-e202-49e6-b18b-53da6a6fbbef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variation_id</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP:247</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP:248</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP:249</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP:251</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953906</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:253</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448523</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15990</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>ClinVar:9570</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448524</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15992</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>ClinVar:690268</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448525</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16002</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:690271</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448526</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16011</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>ClinVar:690273</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448527</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16017</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:690275</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2163505 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene chr       pos ref alt    variation_id variant_type  \\\n",
       "0          LPL   8  19953280   A   A         SNP:247          snv   \n",
       "1          LPL   8  19953315   G   G         SNP:248          snv   \n",
       "2          LPL   8  19953495   T   T         SNP:249          snv   \n",
       "4          LPL   8  19953649   T   T         SNP:251          snv   \n",
       "5          LPL   8  19953906   C   C         SNP:253          snv   \n",
       "...        ...  ..       ...  ..  ..             ...          ...   \n",
       "1448523  MT-TP  MT     15990   C   C    ClinVar:9570          snv   \n",
       "1448524  MT-TP  MT     15992   A   A  ClinVar:690268          snv   \n",
       "1448525  MT-TP  MT     16002   T   T  ClinVar:690271          snv   \n",
       "1448526  MT-TP  MT     16011   A   A  ClinVar:690273          snv   \n",
       "1448527  MT-TP  MT     16017   T   T  ClinVar:690275          snv   \n",
       "\n",
       "        clinical_significance  \n",
       "0                     healthy  \n",
       "1                     healthy  \n",
       "2                     healthy  \n",
       "4                     healthy  \n",
       "5                     healthy  \n",
       "...                       ...  \n",
       "1448523               healthy  \n",
       "1448524               healthy  \n",
       "1448525               healthy  \n",
       "1448526               healthy  \n",
       "1448527               healthy  \n",
       "\n",
       "[2163505 rows x 8 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "healthy_df = final_df.copy()\n",
    "healthy_df = healthy_df[healthy_df['ref'] != \"-\"]\n",
    "healthy_df['alt'] = healthy_df['ref']\n",
    "healthy_df.loc[:, 'clinical_significance'] = ['healthy'] * healthy_df.shape[0]\n",
    "healthy_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "601807ae-8148-4387-aae4-6a7f1dd72d44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variation_id</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953280</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:247</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953315</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP:248</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953495</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:249</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953512</td>\n",
       "      <td>-</td>\n",
       "      <td>TG</td>\n",
       "      <td>SNP:250</td>\n",
       "      <td>delins</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19953649</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP:251</td>\n",
       "      <td>snv</td>\n",
       "      <td>non-harmful</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050448</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15990</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>ClinVar:9570</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050449</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15992</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>ClinVar:690268</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050450</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16002</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:690271</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050451</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16011</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>ClinVar:690273</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050452</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>16017</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>ClinVar:690275</td>\n",
       "      <td>snv</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4050453 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene chr       pos ref alt    variation_id variant_type  \\\n",
       "0          LPL   8  19953280   A   C         SNP:247          snv   \n",
       "1          LPL   8  19953315   G   A         SNP:248          snv   \n",
       "2          LPL   8  19953495   T   C         SNP:249          snv   \n",
       "3          LPL   8  19953512   -  TG         SNP:250       delins   \n",
       "4          LPL   8  19953649   T   C         SNP:251          snv   \n",
       "...        ...  ..       ...  ..  ..             ...          ...   \n",
       "4050448  MT-TP  MT     15990   C   C    ClinVar:9570          snv   \n",
       "4050449  MT-TP  MT     15992   A   A  ClinVar:690268          snv   \n",
       "4050450  MT-TP  MT     16002   T   T  ClinVar:690271          snv   \n",
       "4050451  MT-TP  MT     16011   A   A  ClinVar:690273          snv   \n",
       "4050452  MT-TP  MT     16017   T   T  ClinVar:690275          snv   \n",
       "\n",
       "        clinical_significance  \n",
       "0                 non-harmful  \n",
       "1                 non-harmful  \n",
       "2                 non-harmful  \n",
       "3                 non-harmful  \n",
       "4                 non-harmful  \n",
       "...                       ...  \n",
       "4050448               healthy  \n",
       "4050449               healthy  \n",
       "4050450               healthy  \n",
       "4050451               healthy  \n",
       "4050452               healthy  \n",
       "\n",
       "[4050453 rows x 8 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df = pd.concat([final_df, healthy_df], axis=0)\n",
    "final_df = final_df.drop_duplicates().reset_index().drop(\"index\", axis=1)\n",
    "final_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bb6e95f1-f37b-4dd4-871b-205066e15e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv(\"Mutation_and_healthy_data.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52b83612-946f-4168-846d-5ac0ec131cf7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
